{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Antigenic analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Alphabet import generic_dna\n",
    "import pandas as pd\n",
    "import os\n",
    "import tempfile\n",
    "import sys\n",
    "import subprocess\n",
    "from scipy.stats import binom\n",
    "from ast import literal_eval\n",
    "from ggplot import *\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def write_to_summary(line_pattern,value):\n",
    "    lines = []\n",
    "    with open(\"./results.table.tsv\",'r') as results:\n",
    "        for line in results:\n",
    "            if line_pattern in line:\n",
    "                line = line_pattern +\"\\t\" + str(value)+\"\\n\"\n",
    "            lines.append(line)\n",
    "            \n",
    "    with open(\"./results.table.tsv\",'w') as output:\n",
    "        for l in lines:\n",
    "            output.write(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def Align(headers_seqs, progpath, program='PROBCONS', musclegapopen=None):\n",
    "    \"\"\"Performs a multiple sequence alignment of two or more sequences.\n",
    "\n",
    "    By default, the protein sequences are aligned using PROBCONS.  This is\n",
    "        probably the most accurate alignment program.  However, it is\n",
    "        slow and consumes large amounts of memory if you are aligning\n",
    "        a very large number of sequences (typically if you are aligning\n",
    "        more than several hundred).  In that case, you may prefer to use\n",
    "        MUSCLE instead.  You can choose between the two with the 'program'\n",
    "        option.  If you decide to use MUSCLE, you can also align nucleotide\n",
    "        sequences with this program.\n",
    "\n",
    "    'headers_seqs' is a list specifying the names of the sequences that we\n",
    "        want to align.  Each entry is a 2-tuple '(head, seq)' where 'head' is\n",
    "        a header giving the sequence name and other information (might be empty)\n",
    "        and 'seq' is a string giving the protein sequence.  The list must have\n",
    "        at least 2 entries.\n",
    "\n",
    "    'progpath' should specify a directory containing the alignment program executable,\n",
    "        either PROBCONS or MUSCLE.  The PROBCONS executable is assumed to have\n",
    "        the name \"probcons\" in this directory.  The MUSCLE executable is assumed to\n",
    "        have the name \"muscle\" in this directory.\n",
    "\n",
    "    'program' specifies what program to use for the alignment.  By default, it is\n",
    "        \"PROBCONS\".  If you wish to use MUSCLE instead, set it to \"MUSCLE\".\n",
    "\n",
    "    'musclegapopen' sets the MUSCLE gap openining penalty to the specified\n",
    "        value. By default it is None, meaning we use the MUSCLE default penalty.\n",
    "        You can also set it to a number; for example -100 will lead to fewer gaps.\n",
    "\n",
    "    This executable is used to perform a multiple sequence alignment of the proteins\n",
    "        with the default settings of either PROBCONS or MUSCLE.  The returned variable is a\n",
    "        new list 'aligned_headers_seqs'.  Each entry is a 2-tuple '(head, aligned_seq)'.\n",
    "        'head' has the same meaning as on input (the sequence header) and\n",
    "        'aligned_seq' is the aligned sequence, with gaps inserted as '-'\n",
    "        as appropriate.  Therefore, all of the 'aligned_seq' entries in\n",
    "        'aligned_headers_seqs' are of the same length.  The entries in 'aligned_headers_seq'\n",
    "        are in the same order as in the input list 'headers_seqs'.\n",
    "    \"\"\"\n",
    "    if not (isinstance(headers_seqs, list) and len(headers_seqs) >= 2):\n",
    "        raise ValueError('header_seqs does not specify a list with at least two entries.')\n",
    "    if not os.path.isdir(progpath):\n",
    "        raise ValueError(\"Cannot find directory %s.\" % progpath)\n",
    "    if program == 'PROBCONS':\n",
    "        exe = os.path.abspath(\"%s/probcons\" % progpath) # the executable\n",
    "    elif program == 'MUSCLE':\n",
    "        exe = os.path.abspath(\"%s/muscle\" % progpath) # the executable\n",
    "    else:\n",
    "        raise ValueError(\"Invalid value of %s for 'program'.\" % (str(program)))\n",
    "    if not os.path.isfile(exe):\n",
    "        raise IOError(\"Cannot find executable at %s.\" % exe)\n",
    "    currdir = os.getcwd()\n",
    "    tempdir = tempfile.mkdtemp()\n",
    "    try:\n",
    "        # do stuff in a temporary directory\n",
    "        infile = \"%s/in.fasta\" % tempdir # input file\n",
    "        outfile = \"%s/out.fasta\" % tempdir # output file\n",
    "        WriteFASTA(headers_seqs, infile) # write sequences to the input file\n",
    "        if program == 'PROBCONS':\n",
    "            p = subprocess.Popen(\"%s %s\" % (exe, infile), shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE) # run ProbCons\n",
    "            (output, errors) = p.communicate()\n",
    "\n",
    "            open(outfile, 'w').write(output)\n",
    "        elif program == 'MUSCLE':\n",
    "            if musclegapopen != None:\n",
    "                p = subprocess.Popen(\"%s -gapopen %d -in %s -out %s\" % (exe, musclegapopen, infile, outfile), shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE) # run MUSCLE\n",
    "            else:\n",
    "                p = subprocess.Popen(\"%s -in %s -out %s\" % (exe, infile, outfile), shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE) # run MUSCLE\n",
    "            (output, errors) = p.communicate()\n",
    "        try:\n",
    "            aligned_headers_seqs = ReadFASTA(outfile)\n",
    "        except:\n",
    "            sys.stderr.write(\"Error getting alignment output, error of %s\" % errors)\n",
    "            raise\n",
    "    finally:\n",
    "        os.chdir(currdir) # return to the original directory\n",
    "        for file in os.listdir(tempdir):\n",
    "            os.remove(\"%s/%s\" % (tempdir, file)) # remove files from temporary directory\n",
    "        os.rmdir(tempdir) # remove temporary directory\n",
    "    if len(aligned_headers_seqs) != len(headers_seqs):\n",
    "        raise ValueError(\"Did not return the correct number of aligned sequences.\")\n",
    "    # put the aligned sequences in the same order as the input sequences\n",
    "    n = len(aligned_headers_seqs[0][1]) # length of aligned sequences\n",
    "    d = dict(aligned_headers_seqs)\n",
    "    aligned_headers_seqs = []\n",
    "    for (head, seq) in headers_seqs:\n",
    "        try:\n",
    "            alignedseq = d[head]\n",
    "        except KeyError:\n",
    "            raise ValueError(\"After alignment, the following header is missing: %s\" % head)\n",
    "        if len(alignedseq) != n:\n",
    "            open('errors.temp', 'w').write(errors)\n",
    "            raise ValueError(\"Aligned sequence %s is not of length %d: if you are using MUSCLE, you may be running out of memory.  Errors have been written to errors.temp.\" % (alignedseq, n))\n",
    "        if len(seq) > n:\n",
    "            open('errors.temp', 'w').write(errors)\n",
    "            raise ValueError(\"Unaligned seq %s is longer than aligned length of %d: if you are using MUSCLE, you many be running out of memory.  Errors have been written to errors.temp.\" % (seq, n))\n",
    "        aligned_headers_seqs.append((head, alignedseq))\n",
    "    return aligned_headers_seqs # return the aligned sequences\n",
    "\n",
    "\n",
    "def StripGapsToFirstSequence(aligned_headers_seqs):\n",
    "    \"\"\"Strips gaps from a reference sequence, and all corresponding alignments.\n",
    "\n",
    "    On input, 'aligned_headers_seqs' should be a set of two or more aligned sequences,\n",
    "        as would be returned by 'Align'.\n",
    "\n",
    "    The first sequence in this alignment is taken to correspond to the reference sequence.\n",
    "        The returned variable is a list similar to aligned_headers_seqs, but with\n",
    "        all positions corresponding to gaps in this reference sequence stripped away.\n",
    "        All gaps ('-') characters are removed from this reference sequence.  In addition,\n",
    "        in all other aligned sequences in 'aligned_headers_seqs', every character at\n",
    "        the same position as a gap in the reference sequence is removed.  Therefore,\n",
    "        at the end of this procedure, all of the alignments have the same length\n",
    "        as the reference sequence with its gaps stripped away.  The headers are\n",
    "        unchanged.  The order of sequences in this stripped alignment is also\n",
    "        unchanged.\n",
    "\n",
    "    >>> StripGapsToFirstSequence([('s1', '-AT-A-GC'), ('s2', 'AAT-TAGC'), ('s3', '--T-A-GC')])\n",
    "    [('s1', 'ATAGC'), ('s2', 'ATTGC'), ('s3', '-TAGC')]\n",
    "    \"\"\"\n",
    "    if not (isinstance(aligned_headers_seqs, list) and len(aligned_headers_seqs) >= 2):\n",
    "        raise ValueError(\"aligned_headers_seqs does not specify at least two aligned sequences.\")\n",
    "    (ref_head, ref_seq) = aligned_headers_seqs[0]\n",
    "    non_strip_positions = [] # positions not to strip away\n",
    "    stripped_ref_seq = []\n",
    "    for i in range(len(ref_seq)):\n",
    "        r = ref_seq[i]\n",
    "        if r != '-':\n",
    "            non_strip_positions.append(i)\n",
    "            stripped_ref_seq.append(r)\n",
    "    stripped_headers_seqs = [(ref_head, ''.join(stripped_ref_seq))]\n",
    "    for (ihead, iseq) in aligned_headers_seqs[1 : ]:\n",
    "        istrippedseq = ''.join([iseq[i] for i in non_strip_positions])\n",
    "        stripped_headers_seqs.append((ihead, istrippedseq))\n",
    "    return stripped_headers_seqs\n",
    "\n",
    "\n",
    "def WriteFASTA(headers_seqs, filename, writable_file=False):\n",
    "    \"\"\"Writes sequences to a FASTA file.\n",
    "\n",
    "    'headers_seqs' is a list of 2-tuples specifying sequences and their\n",
    "        corresponding headers.  Each entry is the 2-tuple '(header, seq)'\n",
    "        where 'header' is a string giving the header (without the leading \">\"),\n",
    "        and 'seq' is the corresponding sequence.\n",
    "\n",
    "    'filename' is a string that specifies the name of the file to which the\n",
    "        headers and sequences should be written.  If this file already exists,\n",
    "        it is overwritten.\n",
    "\n",
    "    'writable_file' is a Boolean switch specifying that rather than 'filename'\n",
    "        giving a string specifying the name of a file to which the sequences\n",
    "        should be written, it instead specifies a writable file object to which\n",
    "        the sequences should be written.\n",
    "\n",
    "    The sequences are written to the file in the same order that they are specified\n",
    "        in 'headers_seqs'.\n",
    "    \"\"\"\n",
    "    assert isinstance(writable_file, bool)\n",
    "    if writable_file:\n",
    "        f = filename\n",
    "    else:\n",
    "        f = open(filename, 'w')\n",
    "    for (header, seq) in headers_seqs:\n",
    "        f.write(\">%s\\n%s\\n\" % (header, seq))\n",
    "    if not writable_file:\n",
    "        f.close()\n",
    "\n",
    "\n",
    "\n",
    "def ReadFASTA(fastafile):\n",
    "    \"\"\"Reads sequences from a FASTA file.\n",
    "\n",
    "    'fastafile' should specify the name of a FASTA file.\n",
    "\n",
    "    This function reads all sequences from the FASTA file.  It returns the\n",
    "        list 'headers_seqs'.  This list is composed of a 2-tuple '(header, seq)'\n",
    "        for every sequence entry in FASTA file.  'header' is the header for\n",
    "        a sequence, with the leading \">\" and any trailing spaces removed. 'seq'\n",
    "        is the corresponding sequence.\n",
    "    \"\"\"\n",
    "    lines = open(fastafile).readlines()\n",
    "    headers_seqs = []\n",
    "    header = None\n",
    "    seq = []\n",
    "    for line in lines:\n",
    "        if line[0] == '>':\n",
    "            if (not header) and (not seq):\n",
    "                pass # first sequence in file\n",
    "            elif header and not seq:\n",
    "                raise ValueError(\"Empty sequence for %s\" % header)\n",
    "            elif seq and not header:\n",
    "                raise ValueError(\"File does not begin with header.\")\n",
    "            else:\n",
    "                seq = ''.join(seq)\n",
    "                seq = seq.replace(' ', '')\n",
    "                headers_seqs.append((header, seq))\n",
    "            header = line.strip()[1 : ]\n",
    "            seq = []\n",
    "        else:\n",
    "            seq.append(line.strip())\n",
    "    if (not header) and (not seq):\n",
    "        pass # first sequence in file\n",
    "    elif header and not seq:\n",
    "        raise ValueError(\"Empty sequence for %s\" % header)\n",
    "    elif seq and not header:\n",
    "        raise ValueError(\"File does not begin with header.\")\n",
    "    else:\n",
    "        seq = ''.join(seq)\n",
    "        seq = seq.replace(' ', '')\n",
    "        headers_seqs.append((header, seq))\n",
    "    return headers_seqs\n",
    "\n",
    "\n",
    "def GetCorrespondingResidue(seqs, i):\n",
    "    \"\"\"Gets the corresponding residue number for two aligned sequences.\n",
    "\n",
    "    *seqs* is a set of two aligned sequences as *(head, sequence)* 2-tuples.\n",
    "\n",
    "    *i* is the number of a residue in sequential numbering of *seqs[0]*\n",
    "    without considering any of the gaps induced by alignment, in 1, 2, ...\n",
    "    numbering.\n",
    "\n",
    "    Returns the number of the residue in sequential numbering of *seqs[1]*\n",
    "    without considering any of the gaps induced by alignment in 1, 2, ...\n",
    "    numbering. Returns *None* if residue *i* of *seqs[0]* aligns with a\n",
    "    gap in *seqs[1]*.\n",
    "    \"\"\"\n",
    "    assert len(seqs) == 2\n",
    "    s1 = seqs[0][1]\n",
    "    s2 = seqs[1][1]\n",
    "    assert len(s1) == len(s2)\n",
    "    assert 1 <= i <= len(s1)\n",
    "    s1index = s2index = 0\n",
    "    for j in range(len(s1)):\n",
    "        if s1[j] != '-':\n",
    "            s1index += 1\n",
    "        if s2[j] != '-':\n",
    "            s2index += 1\n",
    "        if s1index == i:\n",
    "            if s2[j] == '-':\n",
    "                return None\n",
    "            else:\n",
    "                return s2index\n",
    "\n",
    "\n",
    "def find_aligments(aligner,aligner_path,seq,pos): # The positions will be given as base 0 and adjusted to match the convention (base 1) in the funciton\n",
    "    \"\"\"Main body of script.\"\"\"\n",
    "    print(\"\\nBeginning execution of HA_numbering script.\")\n",
    "\n",
    "    # parse arguments\n",
    "    if aligner == 'probcons':\n",
    "        alignerpath = aligner_path # redundant\n",
    "        if not os.path.isdir(alignerpath):\n",
    "            raise IOError(\"The directory of %s specified by probconspath does not exist.\" % (alignerpath))\n",
    "        prog = 'PROBCONS'\n",
    "    elif aligner == 'muscle':\n",
    "        alignerpath = aligner_path # redundant\n",
    "        if not os.path.isdir(alignerpath):\n",
    "            raise IOError(\"The directory of %s specified by musclepath does not exist.\" % (alignerpath))\n",
    "        prog = 'MUSCLE'\n",
    "    else:\n",
    "        raise IOError(\"First argument does not specify probcons or muscle\")\n",
    "    ha_sequence = str(seq) # just get the sequence\n",
    "    sites = pos\n",
    "    # Define sequences and their numbering conversions.\n",
    "    # The sequences are in seq_d and keyed by PDB code.\n",
    "    # The number conversions from sequential numbering of these sequences to\n",
    "    # th number labels are in label_d and keyed by PDB codes.\n",
    "    seq_names = ['4HMG', '4JTV']\n",
    "    seq_d = {}\n",
    "    label_d = {}\n",
    "    seq_4hmg_a = \\\n",
    "        'QDLPGNDNSTATLCLGHHAVPNGTLVKTITDDQIEVTNATELVQSSSTGKICNNPHRILDGIDCTLIDALLGDPHCDVFQ' +\\\n",
    "        'NETWDLFVERSKAFSNCYPYDVPDYASLRSLVASSGTLEFITEGFTWTGVTQNGGSNACKRGPGSGFFSRLNWLTKSGST' +\\\n",
    "        'YPVLNVTMPNNDNFDKLYIWGIHHPSTNQEQTSLYVQASGRVTVSTRRSQQTIIPNIGSRPWVRGQSSRISIYWTIVKPG' +\\\n",
    "        'DVLVINSNGNLIAPRGYFKMRTGKSSIMRSDAPIDTCISECITPNGSIPNDKPFQNVNKITYGACPKYVKQNTLKLATGM' +\\\n",
    "        'RNVPEKQT'\n",
    "    seq_4hmg_b = \\\n",
    "        'GLFGAIAGFIENGWEGMIDGWYGFRHQNSEGTGQAADLKSTQAAIDQINGKLNRVIEKTNEKFHQIEKEFSEVEGRIQDL' +\\\n",
    "        'EKYVEDTKIDLWSYNAELLVALENQHTIDLTDSEMNKLFEKTRRQLRENAEEMGNGCFKIYHKCDNACIESIRNGTYDHD' +\\\n",
    "        'VYRDEALNNRFQIKG'\n",
    "    seq_d['4HMG'] = seq_4hmg_a + seq_4hmg_b\n",
    "    label_d['4HMG'] = dict([(i + 1, '%d in HA1' % (i + 1)) for i in range(len(seq_4hmg_a))] + [(len(seq_4hmg_a) + i + 1, '%d in HA2' % (i + 1)) for i in range(len(seq_4hmg_b))])\n",
    "    assert len(seq_d['4HMG']) == len(label_d['4HMG'])\n",
    "    seq_4jtv_a = \\\n",
    "        'DTLCIGYHANNSTDTVDTVLEKNVTVTHSVNLLEDKHNGKLCKLRGVAPLHLGKCNIAGWILGNPECESLSTASSWSYIV' +\\\n",
    "        'ETPSSDNGTCYPGDFIDYEELREQLSSVSSFERFEIFPKTSSWPNHDSNKGVTAACPHAGAKSFYKNLIWLVKKGNSYPK' +\\\n",
    "        'LSKSYINDKGKEVLVLWGIHHPSTSADQQSLYQNADTYVFVGSSRYSKKFKPEIAIRPKVRDQEGRMNYYWTLVEPGDKI' +\\\n",
    "        'TFEATGNLVVPRYAFAMERNAGSGIIISDTPVHDCNTTCQTPKGAINTSLPFQNIHPITIGKCPKYVKSTKLRLATGLRN' +\\\n",
    "        'I'\n",
    "    seq_4jtv_b = \\\n",
    "        'GLFGAIAGFIEGGWTGMVDGWYGYHHQNEQGSGYAADLKSTQNAIDEITNKVNSVIEKMNTQFTAVGKEFNHLEKRIENL' +\\\n",
    "        'NKKVDDGFLDIWTYNAELLVLLENERTLDYHDSNVKNLYEKVRSQLKNNAKEIGNGCFEFYHKCDNTCMESVKNGTYDYP' +\\\n",
    "        'KY'\n",
    "    seq_d['4JTV'] = seq_4jtv_a + seq_4jtv_b\n",
    "    label_d['4JTV'] = dict([(i + 1, '%d of HA1' % (i + 7)) for i in range(len(seq_4jtv_a))] + [(len(seq_4jtv_a) + i + 1, '%d of HA2' % (i + 1)) for i in range(len(seq_4jtv_b))])\n",
    "    assert len(seq_d['4JTV']) == len(label_d['4JTV'])\n",
    "\n",
    "    # make alignments\n",
    "    print(\"Making %s alignments...\" % prog)\n",
    "    alignments = {}\n",
    "    for seqname in seq_names:\n",
    "        alignments[seqname] = Align([('seq', ha_sequence), (seqname, seq_d[seqname])], alignerpath, prog)\n",
    "    print(\"Alignments complete.\\n\\nHere are the corresponding residue numbers:\")\n",
    "    #print(sites)\n",
    "    out_hmg=[]\n",
    "    out_jtv = []\n",
    "    for site in sites:\n",
    "        #site=site+1 # to correct for difference in base numbering\n",
    "        if not (1 <= site <= len(ha_sequence)):\n",
    "            raise ValueError(\"site %d is outside the valid range for sequential numbering of ha_sequence starting at 1.\" % site)\n",
    "        sitestring = ['\\nResidue %s%d in sequential numbering of your HA sequence corresponds to:' % (ha_sequence[site - 1], site)]\n",
    "        for seqname in seq_names:\n",
    "            i = GetCorrespondingResidue(alignments[seqname], site)\n",
    "            if i == None:\n",
    "                sitestring.append('  * an alignment gap in %s' % seqname)\n",
    "            else:\n",
    "                if not (1 <= i <= len(seq_d[seqname])):\n",
    "                    raise ValueError(\"Invalid corrresponding residue for %s -- something is wrong with this program\" % seqname)\n",
    "                aa = seq_d[seqname][i - 1]\n",
    "\n",
    "                sitestring.append('  * %s%s in %s' % (aa, label_d[seqname][i], seqname))\n",
    "        print( '\\n'.join(sitestring))\n",
    "        out_hmg.append(sitestring[1][4:-7])\n",
    "        out_jtv.append(sitestring[2][4:-7])\n",
    "    \n",
    "    return([out_hmg,out_jtv])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting the  antigenic regions\n",
    "\n",
    "## H1N1\n",
    "Here I have copied the antigenic sites listed in Xu et al. 2010. I am using a derivation of the Bloom lab script to align them to the PBD protien. The script requires base 1 numbering\n",
    "## H3N2\n",
    "I am also reading in H3 antigenic regions noted in Lee et al. 2004 with a special column noting if the site was also mentioned in Wiley et al. 1981 and/or Smith et al. 2004.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Antigenic site strings  \"site : [ [list of sequences], [List of lists of antigenic positions in that sequence]]\n",
    "H1N1 = {'Sa' : [[\"PNHDS\",\"KGNSYPKLSKS\"],[[1,2],[1,2,3,4,6,7,8,9,10,11]]] ,\n",
    "        'Sb':[[\"TSADQQSLYQNA\"],[[1,2,3,4,5,6,7,8,9,10,11,12]]],\n",
    "        'Ca' : [[\"PHAGAKSF\",\"INDKGK\",\"SSRYSK\",\"KVRDQ\",\"EPGDKI\"],[[1,2,3,4,5,6],[1,2,3,4,5],[1,2,3],[3,4],[1,2,3]]],\n",
    "        'Cb' : [[\"LSTASSWSY\"], [[1,2,3,4,5,6]]]}\n",
    "\n",
    "H3 = pd.read_csv(\"../data/reference/H3N2_antigenic_sites.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This chunck does the H1N1 aligments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TSADQQSLYQNA\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T187 in HA1 in 4HMG\n",
      "  * T190 of HA1 in 4JTV\n",
      "\n",
      "Residue S2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N188 in HA1 in 4HMG\n",
      "  * S191 of HA1 in 4JTV\n",
      "\n",
      "Residue A3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Q189 in HA1 in 4HMG\n",
      "  * A192 of HA1 in 4JTV\n",
      "\n",
      "Residue D4 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E190 in HA1 in 4HMG\n",
      "  * D193 of HA1 in 4JTV\n",
      "\n",
      "Residue Q5 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Q191 in HA1 in 4HMG\n",
      "  * Q194 of HA1 in 4JTV\n",
      "\n",
      "Residue Q6 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T192 in HA1 in 4HMG\n",
      "  * Q195 of HA1 in 4JTV\n",
      "\n",
      "Residue S7 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S193 in HA1 in 4HMG\n",
      "  * S196 of HA1 in 4JTV\n",
      "\n",
      "Residue L8 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L194 in HA1 in 4HMG\n",
      "  * L197 of HA1 in 4JTV\n",
      "\n",
      "Residue Y9 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Y195 in HA1 in 4HMG\n",
      "  * Y198 of HA1 in 4JTV\n",
      "\n",
      "Residue Q10 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V196 in HA1 in 4HMG\n",
      "  * Q199 of HA1 in 4JTV\n",
      "\n",
      "Residue N11 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Q197 in HA1 in 4HMG\n",
      "  * N200 of HA1 in 4JTV\n",
      "\n",
      "Residue A12 in sequential numbering of your HA sequence corresponds to:\n",
      "  * A198 in HA1 in 4HMG\n",
      "  * A201 of HA1 in 4JTV\n",
      "LSTASSWSY\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L154 in HA1 in 4HMG\n",
      "  * L76 of HA1 in 4JTV\n",
      "\n",
      "Residue S2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T155 in HA1 in 4HMG\n",
      "  * S77 of HA1 in 4JTV\n",
      "\n",
      "Residue T3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K156 in HA1 in 4HMG\n",
      "  * T78 of HA1 in 4JTV\n",
      "\n",
      "Residue A4 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S157 in HA1 in 4HMG\n",
      "  * A79 of HA1 in 4JTV\n",
      "\n",
      "Residue S5 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G158 in HA1 in 4HMG\n",
      "  * S80 of HA1 in 4JTV\n",
      "\n",
      "Residue S6 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S159 in HA1 in 4HMG\n",
      "  * S81 of HA1 in 4JTV\n",
      "PHAGAKSF\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P215 in HA1 in 4HMG\n",
      "  * P143 of HA1 in 4JTV\n",
      "\n",
      "Residue H2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N216 in HA1 in 4HMG\n",
      "  * H144 of HA1 in 4JTV\n",
      "\n",
      "Residue A3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * I217 in HA1 in 4HMG\n",
      "  * A145 of HA1 in 4JTV\n",
      "\n",
      "Residue G4 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G218 in HA1 in 4HMG\n",
      "  * G146 of HA1 in 4JTV\n",
      "\n",
      "Residue A5 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S219 in HA1 in 4HMG\n",
      "  * A147 of HA1 in 4JTV\n",
      "\n",
      "Residue K6 in sequential numbering of your HA sequence corresponds to:\n",
      "  * R220 in HA1 in 4HMG\n",
      "  * K148 of HA1 in 4JTV\n",
      "INDKGK\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * I245 in HA1 in 4HMG\n",
      "  * I172 of HA1 in 4JTV\n",
      "\n",
      "Residue N2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N246 in HA1 in 4HMG\n",
      "  * N173 of HA1 in 4JTV\n",
      "\n",
      "Residue D3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S247 in HA1 in 4HMG\n",
      "  * D174 of HA1 in 4JTV\n",
      "\n",
      "Residue K4 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N248 in HA1 in 4HMG\n",
      "  * K175 of HA1 in 4JTV\n",
      "\n",
      "Residue G5 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G249 in HA1 in 4HMG\n",
      "  * G176 of HA1 in 4JTV\n",
      "SSRYSK\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * C305 in HA1 in 4HMG\n",
      "  * S209 of HA1 in 4JTV\n",
      "\n",
      "Residue S2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P306 in HA1 in 4HMG\n",
      "  * S210 of HA1 in 4JTV\n",
      "\n",
      "Residue R3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K307 in HA1 in 4HMG\n",
      "  * R211 of HA1 in 4JTV\n",
      "KVRDQ\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * R127 in HA2 in 4HMG\n",
      "  * R227 of HA1 in 4JTV\n",
      "\n",
      "Residue D4 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E128 in HA2 in 4HMG\n",
      "  * D228 of HA1 in 4JTV\n",
      "EPGDKI\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K238 in HA1 in 4HMG\n",
      "  * E241 of HA1 in 4JTV\n",
      "\n",
      "Residue P2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P239 in HA1 in 4HMG\n",
      "  * P242 of HA1 in 4JTV\n",
      "\n",
      "Residue G3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G240 in HA1 in 4HMG\n",
      "  * G243 of HA1 in 4JTV\n",
      "PNHDS\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P169 in HA1 in 4HMG\n",
      "  * P130 of HA1 in 4JTV\n",
      "\n",
      "Residue N2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N170 in HA1 in 4HMG\n",
      "  * N131 of HA1 in 4JTV\n",
      "KGNSYPKLSKS\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S157 in HA1 in 4HMG\n",
      "  * K160 of HA1 in 4JTV\n",
      "\n",
      "Residue G2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G158 in HA1 in 4HMG\n",
      "  * G161 of HA1 in 4JTV\n",
      "\n",
      "Residue N3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S159 in HA1 in 4HMG\n",
      "  * N162 of HA1 in 4JTV\n",
      "\n",
      "Residue S4 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T160 in HA1 in 4HMG\n",
      "  * S163 of HA1 in 4JTV\n",
      "\n",
      "Residue P6 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P162 in HA1 in 4HMG\n",
      "  * P165 of HA1 in 4JTV\n",
      "\n",
      "Residue K7 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V163 in HA1 in 4HMG\n",
      "  * K166 of HA1 in 4JTV\n",
      "\n",
      "Residue L8 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L164 in HA1 in 4HMG\n",
      "  * L167 of HA1 in 4JTV\n",
      "\n",
      "Residue S9 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N165 in HA1 in 4HMG\n",
      "  * S168 of HA1 in 4JTV\n",
      "\n",
      "Residue K10 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V166 in HA1 in 4HMG\n",
      "  * K169 of HA1 in 4JTV\n",
      "\n",
      "Residue S11 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T167 in HA1 in 4HMG\n",
      "  * S170 of HA1 in 4JTV\n",
      "['T190 of HA1 ', 'S191 of HA1 ', 'A192 of HA1 ', 'D193 of HA1 ', 'Q194 of HA1 ', 'Q195 of HA1 ', 'S196 of HA1 ', 'L197 of HA1 ', 'Y198 of HA1 ', 'Q199 of HA1 ', 'N200 of HA1 ', 'A201 of HA1 ', 'L76 of HA1 ', 'S77 of HA1 ', 'T78 of HA1 ', 'A79 of HA1 ', 'S80 of HA1 ', 'S81 of HA1 ', 'P143 of HA1 ', 'H144 of HA1 ', 'A145 of HA1 ', 'G146 of HA1 ', 'A147 of HA1 ', 'K148 of HA1 ', 'I172 of HA1 ', 'N173 of HA1 ', 'D174 of HA1 ', 'K175 of HA1 ', 'G176 of HA1 ', 'S209 of HA1 ', 'S210 of HA1 ', 'R211 of HA1 ', 'R227 of HA1 ', 'D228 of HA1 ', 'E241 of HA1 ', 'P242 of HA1 ', 'G243 of HA1 ', 'P130 of HA1 ', 'N131 of HA1 ', 'K160 of HA1 ', 'G161 of HA1 ', 'N162 of HA1 ', 'S163 of HA1 ', 'P165 of HA1 ', 'K166 of HA1 ', 'L167 of HA1 ', 'S168 of HA1 ', 'K169 of HA1 ', 'S170 of HA1 ']\n"
     ]
    }
   ],
   "source": [
    "antigenic = []\n",
    "for site in H1N1:\n",
    "    i = 0\n",
    "    for seq in H1N1[site][0]:\n",
    "        print(seq)\n",
    "        nice = find_aligments('muscle',\"/Users/jt/muscle3.8.31/\",seq,H1N1[site][1][i])\n",
    "        for pos in nice[1]:\n",
    "            antigenic.append(pos)\n",
    "        i+=1\n",
    "\n",
    "print(antigenic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Anitgenic analysis\n",
    "\n",
    "Here we will be reading in the quality snv from our analysis. We will focus on Nonsynonymous variants in HA for obvious reasons And I am selecting only minor variants.\n",
    "\n",
    "Note : Here the reference amino acid is the sample consensus not the plasmid control (as is the case for the nucleotide)\n",
    "\n",
    "For each ciruculating strain we will for each variant we will find the 4HMG and 4JTV sequence location. These will be used to identify variants in antigenic regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2723: DtypeWarning: Columns (35) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "snv = pd.read_csv(\"/Users/jt/Documents/Analysis/HIVE/data/processed/qual.snv.csv\",converters={\"AA_pos\": literal_eval})\n",
    "snv = snv.loc[snv['freq.var']<0.5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>X</th>\n",
       "      <th>Unnamed..0.1</th>\n",
       "      <th>Id</th>\n",
       "      <th>MapQ</th>\n",
       "      <th>Phred</th>\n",
       "      <th>Read_pos</th>\n",
       "      <th>chr</th>\n",
       "      <th>cov.ctrl.bw</th>\n",
       "      <th>cov.ctrl.fw</th>\n",
       "      <th>...</th>\n",
       "      <th>DPI</th>\n",
       "      <th>season</th>\n",
       "      <th>log_copy_num</th>\n",
       "      <th>gc_ul</th>\n",
       "      <th>HIGHSD</th>\n",
       "      <th>sequenced</th>\n",
       "      <th>home_collected</th>\n",
       "      <th>class_factor</th>\n",
       "      <th>control.freq</th>\n",
       "      <th>exp.freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>4917</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>1168.0</td>\n",
       "      <td>41.533913</td>\n",
       "      <td>37.458841</td>\n",
       "      <td>61.382029</td>\n",
       "      <td>NR</td>\n",
       "      <td>18013</td>\n",
       "      <td>18212</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>10-11</td>\n",
       "      <td>6.136662</td>\n",
       "      <td>97843.843292</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000193</td>\n",
       "      <td>0.063075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>7250</td>\n",
       "      <td>2</td>\n",
       "      <td>115</td>\n",
       "      <td>1168.0</td>\n",
       "      <td>41.674238</td>\n",
       "      <td>37.528392</td>\n",
       "      <td>59.034668</td>\n",
       "      <td>PA</td>\n",
       "      <td>15825</td>\n",
       "      <td>17751</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>10-11</td>\n",
       "      <td>6.136662</td>\n",
       "      <td>97843.843292</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.045823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>7753</td>\n",
       "      <td>4</td>\n",
       "      <td>122</td>\n",
       "      <td>1168.0</td>\n",
       "      <td>41.498099</td>\n",
       "      <td>37.505703</td>\n",
       "      <td>62.321466</td>\n",
       "      <td>PA</td>\n",
       "      <td>18084</td>\n",
       "      <td>17365</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>10-11</td>\n",
       "      <td>6.136662</td>\n",
       "      <td>97843.843292</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.067699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>9660</td>\n",
       "      <td>5</td>\n",
       "      <td>169</td>\n",
       "      <td>1168.0</td>\n",
       "      <td>41.594608</td>\n",
       "      <td>37.274020</td>\n",
       "      <td>63.987255</td>\n",
       "      <td>PB1</td>\n",
       "      <td>27047</td>\n",
       "      <td>23837</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>10-11</td>\n",
       "      <td>6.136662</td>\n",
       "      <td>97843.843292</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.045392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>9709</td>\n",
       "      <td>6</td>\n",
       "      <td>170</td>\n",
       "      <td>1168.0</td>\n",
       "      <td>41.669893</td>\n",
       "      <td>37.404982</td>\n",
       "      <td>62.524217</td>\n",
       "      <td>PB1</td>\n",
       "      <td>25999</td>\n",
       "      <td>27035</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>10-11</td>\n",
       "      <td>6.136662</td>\n",
       "      <td>97843.843292</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000170</td>\n",
       "      <td>0.462501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>12861</td>\n",
       "      <td>13493</td>\n",
       "      <td>241</td>\n",
       "      <td>1168.0</td>\n",
       "      <td>41.618057</td>\n",
       "      <td>37.460768</td>\n",
       "      <td>64.205162</td>\n",
       "      <td>PB2</td>\n",
       "      <td>25826</td>\n",
       "      <td>24738</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>10-11</td>\n",
       "      <td>6.136662</td>\n",
       "      <td>97843.843292</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.445909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>14022</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>1121.0</td>\n",
       "      <td>38.713415</td>\n",
       "      <td>37.501220</td>\n",
       "      <td>62.269512</td>\n",
       "      <td>HA</td>\n",
       "      <td>17196</td>\n",
       "      <td>19477</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.060653</td>\n",
       "      <td>8213.445551</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.020245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>21902</td>\n",
       "      <td>13452</td>\n",
       "      <td>111</td>\n",
       "      <td>1121.0</td>\n",
       "      <td>41.534214</td>\n",
       "      <td>37.391014</td>\n",
       "      <td>65.533185</td>\n",
       "      <td>PA</td>\n",
       "      <td>18262</td>\n",
       "      <td>11558</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.060653</td>\n",
       "      <td>8213.445551</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.260802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>30573</td>\n",
       "      <td>11</td>\n",
       "      <td>60</td>\n",
       "      <td>1217.0</td>\n",
       "      <td>41.729167</td>\n",
       "      <td>37.629960</td>\n",
       "      <td>64.272817</td>\n",
       "      <td>NP</td>\n",
       "      <td>30202</td>\n",
       "      <td>29646</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>11-12</td>\n",
       "      <td>4.240458</td>\n",
       "      <td>1242.597283</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000184</td>\n",
       "      <td>0.026508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>31635</td>\n",
       "      <td>7</td>\n",
       "      <td>57</td>\n",
       "      <td>1217.0</td>\n",
       "      <td>40.702510</td>\n",
       "      <td>37.346636</td>\n",
       "      <td>59.452422</td>\n",
       "      <td>NR</td>\n",
       "      <td>8118</td>\n",
       "      <td>12397</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>11-12</td>\n",
       "      <td>4.240458</td>\n",
       "      <td>1242.597283</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000146</td>\n",
       "      <td>0.269504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>32073</td>\n",
       "      <td>8</td>\n",
       "      <td>62</td>\n",
       "      <td>1217.0</td>\n",
       "      <td>40.052882</td>\n",
       "      <td>37.469222</td>\n",
       "      <td>61.655568</td>\n",
       "      <td>NR</td>\n",
       "      <td>12759</td>\n",
       "      <td>12972</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>11-12</td>\n",
       "      <td>4.240458</td>\n",
       "      <td>1242.597283</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000272</td>\n",
       "      <td>0.137748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>37668</td>\n",
       "      <td>32</td>\n",
       "      <td>175</td>\n",
       "      <td>1217.0</td>\n",
       "      <td>41.689053</td>\n",
       "      <td>37.503694</td>\n",
       "      <td>64.870719</td>\n",
       "      <td>PB1</td>\n",
       "      <td>26113</td>\n",
       "      <td>19314</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>11-12</td>\n",
       "      <td>4.240458</td>\n",
       "      <td>1242.597283</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000286</td>\n",
       "      <td>0.131039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588</th>\n",
       "      <td>39151</td>\n",
       "      <td>36</td>\n",
       "      <td>198</td>\n",
       "      <td>1217.0</td>\n",
       "      <td>39.587903</td>\n",
       "      <td>37.444355</td>\n",
       "      <td>60.845161</td>\n",
       "      <td>PB2</td>\n",
       "      <td>23892</td>\n",
       "      <td>26738</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>11-12</td>\n",
       "      <td>4.240458</td>\n",
       "      <td>1242.597283</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000138</td>\n",
       "      <td>0.045310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713</th>\n",
       "      <td>49574</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "      <td>1150.0</td>\n",
       "      <td>41.491339</td>\n",
       "      <td>37.338583</td>\n",
       "      <td>58.414173</td>\n",
       "      <td>PB1</td>\n",
       "      <td>10730</td>\n",
       "      <td>19197</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.922566</td>\n",
       "      <td>59763.735122</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.035542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>49578</td>\n",
       "      <td>2</td>\n",
       "      <td>110</td>\n",
       "      <td>1150.0</td>\n",
       "      <td>41.744541</td>\n",
       "      <td>37.207424</td>\n",
       "      <td>57.943231</td>\n",
       "      <td>PB1</td>\n",
       "      <td>10774</td>\n",
       "      <td>18512</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.922566</td>\n",
       "      <td>59763.735122</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000137</td>\n",
       "      <td>0.026382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>857</th>\n",
       "      <td>59799</td>\n",
       "      <td>13448</td>\n",
       "      <td>69</td>\n",
       "      <td>1213.0</td>\n",
       "      <td>36.866667</td>\n",
       "      <td>37.377778</td>\n",
       "      <td>59.311111</td>\n",
       "      <td>NS</td>\n",
       "      <td>7369</td>\n",
       "      <td>24610</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>11-12</td>\n",
       "      <td>5.542293</td>\n",
       "      <td>24898.041190</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.022819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1017</th>\n",
       "      <td>74612</td>\n",
       "      <td>2</td>\n",
       "      <td>79</td>\n",
       "      <td>1125.0</td>\n",
       "      <td>41.608527</td>\n",
       "      <td>37.136628</td>\n",
       "      <td>61.093346</td>\n",
       "      <td>PA</td>\n",
       "      <td>10743</td>\n",
       "      <td>11016</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>10-11</td>\n",
       "      <td>4.851537</td>\n",
       "      <td>5074.685473</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000322</td>\n",
       "      <td>0.210355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1066</th>\n",
       "      <td>76725</td>\n",
       "      <td>3</td>\n",
       "      <td>103</td>\n",
       "      <td>1125.0</td>\n",
       "      <td>41.486762</td>\n",
       "      <td>37.289206</td>\n",
       "      <td>63.089613</td>\n",
       "      <td>PB1</td>\n",
       "      <td>18948</td>\n",
       "      <td>19386</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>10-11</td>\n",
       "      <td>4.851537</td>\n",
       "      <td>5074.685473</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.028421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1328</th>\n",
       "      <td>94524</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>1179.0</td>\n",
       "      <td>39.365166</td>\n",
       "      <td>37.465611</td>\n",
       "      <td>56.783276</td>\n",
       "      <td>HA</td>\n",
       "      <td>3281</td>\n",
       "      <td>10176</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.621547</td>\n",
       "      <td>29882.647245</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000297</td>\n",
       "      <td>0.071512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1360</th>\n",
       "      <td>96618</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>1179.0</td>\n",
       "      <td>40.964374</td>\n",
       "      <td>37.372825</td>\n",
       "      <td>63.579950</td>\n",
       "      <td>M</td>\n",
       "      <td>15210</td>\n",
       "      <td>12070</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.621547</td>\n",
       "      <td>29882.647245</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.028234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1476</th>\n",
       "      <td>105503</td>\n",
       "      <td>4</td>\n",
       "      <td>154</td>\n",
       "      <td>1179.0</td>\n",
       "      <td>39.769706</td>\n",
       "      <td>37.250386</td>\n",
       "      <td>32.769706</td>\n",
       "      <td>PB2</td>\n",
       "      <td>498</td>\n",
       "      <td>2409</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.621547</td>\n",
       "      <td>29882.647245</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000688</td>\n",
       "      <td>0.214451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1493</th>\n",
       "      <td>106090</td>\n",
       "      <td>5</td>\n",
       "      <td>171</td>\n",
       "      <td>1179.0</td>\n",
       "      <td>41.679124</td>\n",
       "      <td>37.328608</td>\n",
       "      <td>63.149485</td>\n",
       "      <td>PB2</td>\n",
       "      <td>18448</td>\n",
       "      <td>18166</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.621547</td>\n",
       "      <td>29882.647245</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000137</td>\n",
       "      <td>0.021868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1640</th>\n",
       "      <td>116682</td>\n",
       "      <td>1</td>\n",
       "      <td>108</td>\n",
       "      <td>1154.0</td>\n",
       "      <td>39.007712</td>\n",
       "      <td>37.182519</td>\n",
       "      <td>33.053985</td>\n",
       "      <td>PB1</td>\n",
       "      <td>1195</td>\n",
       "      <td>5654</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>10-11</td>\n",
       "      <td>4.931225</td>\n",
       "      <td>6096.725833</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000292</td>\n",
       "      <td>0.196564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2206</th>\n",
       "      <td>148508</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>1139.0</td>\n",
       "      <td>39.331133</td>\n",
       "      <td>37.403740</td>\n",
       "      <td>57.577558</td>\n",
       "      <td>HA</td>\n",
       "      <td>5812</td>\n",
       "      <td>12753</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.288065</td>\n",
       "      <td>13865.536380</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>0.032483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2242</th>\n",
       "      <td>151784</td>\n",
       "      <td>1</td>\n",
       "      <td>45</td>\n",
       "      <td>1139.0</td>\n",
       "      <td>41.789398</td>\n",
       "      <td>37.477077</td>\n",
       "      <td>60.324737</td>\n",
       "      <td>NP</td>\n",
       "      <td>21150</td>\n",
       "      <td>22814</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.288065</td>\n",
       "      <td>13865.536380</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.056103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2351</th>\n",
       "      <td>158060</td>\n",
       "      <td>2</td>\n",
       "      <td>147</td>\n",
       "      <td>1139.0</td>\n",
       "      <td>41.801527</td>\n",
       "      <td>37.264122</td>\n",
       "      <td>63.155725</td>\n",
       "      <td>PB1</td>\n",
       "      <td>19342</td>\n",
       "      <td>19402</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.288065</td>\n",
       "      <td>13865.536380</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000310</td>\n",
       "      <td>0.024857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2490</th>\n",
       "      <td>163867</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>1174.0</td>\n",
       "      <td>41.252910</td>\n",
       "      <td>37.413757</td>\n",
       "      <td>62.187302</td>\n",
       "      <td>M</td>\n",
       "      <td>14105</td>\n",
       "      <td>15148</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.318185</td>\n",
       "      <td>14861.296029</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.021896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2507</th>\n",
       "      <td>164821</td>\n",
       "      <td>2</td>\n",
       "      <td>53</td>\n",
       "      <td>1174.0</td>\n",
       "      <td>41.427141</td>\n",
       "      <td>37.454950</td>\n",
       "      <td>59.296997</td>\n",
       "      <td>NP</td>\n",
       "      <td>13767</td>\n",
       "      <td>21388</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.318185</td>\n",
       "      <td>14861.296029</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000398</td>\n",
       "      <td>0.025191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2650</th>\n",
       "      <td>171737</td>\n",
       "      <td>3</td>\n",
       "      <td>200</td>\n",
       "      <td>1174.0</td>\n",
       "      <td>41.652006</td>\n",
       "      <td>37.199846</td>\n",
       "      <td>60.466821</td>\n",
       "      <td>PB1</td>\n",
       "      <td>18770</td>\n",
       "      <td>20262</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>10-11</td>\n",
       "      <td>5.318185</td>\n",
       "      <td>14861.296029</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>0.038458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3157</th>\n",
       "      <td>200805</td>\n",
       "      <td>4</td>\n",
       "      <td>228</td>\n",
       "      <td>1162.0</td>\n",
       "      <td>41.762020</td>\n",
       "      <td>37.545410</td>\n",
       "      <td>61.382224</td>\n",
       "      <td>PB2</td>\n",
       "      <td>25812</td>\n",
       "      <td>25609</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>10-11</td>\n",
       "      <td>4.812868</td>\n",
       "      <td>4642.371815</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000136</td>\n",
       "      <td>0.047676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30352</th>\n",
       "      <td>9489147</td>\n",
       "      <td>7</td>\n",
       "      <td>85</td>\n",
       "      <td>MH8692_A</td>\n",
       "      <td>41.063368</td>\n",
       "      <td>36.944469</td>\n",
       "      <td>60.461039</td>\n",
       "      <td>PB1</td>\n",
       "      <td>10798</td>\n",
       "      <td>11046</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>4.418474</td>\n",
       "      <td>1872.171013</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000229</td>\n",
       "      <td>0.174161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30378</th>\n",
       "      <td>12705147</td>\n",
       "      <td>10</td>\n",
       "      <td>113</td>\n",
       "      <td>MH8692_A</td>\n",
       "      <td>41.463320</td>\n",
       "      <td>37.500386</td>\n",
       "      <td>63.997683</td>\n",
       "      <td>PB2</td>\n",
       "      <td>11916</td>\n",
       "      <td>13281</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>4.418474</td>\n",
       "      <td>1872.171013</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000317</td>\n",
       "      <td>0.041392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30385</th>\n",
       "      <td>13024147</td>\n",
       "      <td>11</td>\n",
       "      <td>119</td>\n",
       "      <td>MH8692_A</td>\n",
       "      <td>41.003210</td>\n",
       "      <td>37.369181</td>\n",
       "      <td>63.711075</td>\n",
       "      <td>PB2</td>\n",
       "      <td>12266</td>\n",
       "      <td>10707</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>4.418474</td>\n",
       "      <td>1872.171013</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.022652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30452</th>\n",
       "      <td>8229148</td>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>MH8700_A</td>\n",
       "      <td>41.106218</td>\n",
       "      <td>37.446891</td>\n",
       "      <td>63.032383</td>\n",
       "      <td>PA</td>\n",
       "      <td>8186</td>\n",
       "      <td>7839</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>5.462005</td>\n",
       "      <td>20695.554297</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000187</td>\n",
       "      <td>0.024680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30508</th>\n",
       "      <td>10011029</td>\n",
       "      <td>13558</td>\n",
       "      <td>8</td>\n",
       "      <td>MH9045_A</td>\n",
       "      <td>41.616683</td>\n",
       "      <td>37.487918</td>\n",
       "      <td>63.541377</td>\n",
       "      <td>HA</td>\n",
       "      <td>11606</td>\n",
       "      <td>12259</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>5.285284</td>\n",
       "      <td>13777.027742</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.160589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30576</th>\n",
       "      <td>7594149</td>\n",
       "      <td>2</td>\n",
       "      <td>106</td>\n",
       "      <td>MH9045_A</td>\n",
       "      <td>41.651429</td>\n",
       "      <td>37.492381</td>\n",
       "      <td>62.947619</td>\n",
       "      <td>PA</td>\n",
       "      <td>8430</td>\n",
       "      <td>7612</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>5.285284</td>\n",
       "      <td>13777.027742</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000187</td>\n",
       "      <td>0.036735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30647</th>\n",
       "      <td>12323152</td>\n",
       "      <td>7</td>\n",
       "      <td>229</td>\n",
       "      <td>MH9045_A</td>\n",
       "      <td>41.553222</td>\n",
       "      <td>37.402607</td>\n",
       "      <td>60.401159</td>\n",
       "      <td>PB2</td>\n",
       "      <td>10801</td>\n",
       "      <td>12104</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>5.285284</td>\n",
       "      <td>13777.027742</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.059973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30712</th>\n",
       "      <td>4945150</td>\n",
       "      <td>3</td>\n",
       "      <td>39</td>\n",
       "      <td>MH8925_A</td>\n",
       "      <td>41.588921</td>\n",
       "      <td>37.536443</td>\n",
       "      <td>61.093294</td>\n",
       "      <td>NR</td>\n",
       "      <td>1914</td>\n",
       "      <td>1892</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>5.684998</td>\n",
       "      <td>34583.562088</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30725</th>\n",
       "      <td>6241190</td>\n",
       "      <td>4</td>\n",
       "      <td>52</td>\n",
       "      <td>MH8925_A</td>\n",
       "      <td>41.602174</td>\n",
       "      <td>37.639130</td>\n",
       "      <td>61.995652</td>\n",
       "      <td>NS</td>\n",
       "      <td>18029</td>\n",
       "      <td>17069</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>5.684998</td>\n",
       "      <td>34583.562088</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000342</td>\n",
       "      <td>0.027553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30780</th>\n",
       "      <td>11021190</td>\n",
       "      <td>7</td>\n",
       "      <td>122</td>\n",
       "      <td>MH8925_A</td>\n",
       "      <td>40.516971</td>\n",
       "      <td>37.501305</td>\n",
       "      <td>75.872063</td>\n",
       "      <td>PB1</td>\n",
       "      <td>6384</td>\n",
       "      <td>1206</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>5.684998</td>\n",
       "      <td>34583.562088</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000791</td>\n",
       "      <td>0.044519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30789</th>\n",
       "      <td>12127150</td>\n",
       "      <td>8</td>\n",
       "      <td>132</td>\n",
       "      <td>MH8925_A</td>\n",
       "      <td>41.527886</td>\n",
       "      <td>37.487678</td>\n",
       "      <td>61.630350</td>\n",
       "      <td>PB2</td>\n",
       "      <td>11121</td>\n",
       "      <td>11413</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>5.684998</td>\n",
       "      <td>34583.562088</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>0.029041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30855</th>\n",
       "      <td>7280151</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>MH8924_B</td>\n",
       "      <td>41.671346</td>\n",
       "      <td>37.517754</td>\n",
       "      <td>64.987614</td>\n",
       "      <td>PA</td>\n",
       "      <td>7607</td>\n",
       "      <td>7617</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>5.064007</td>\n",
       "      <td>8277.116060</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000263</td>\n",
       "      <td>0.057796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30882</th>\n",
       "      <td>10263154</td>\n",
       "      <td>1</td>\n",
       "      <td>90</td>\n",
       "      <td>MH8924_B</td>\n",
       "      <td>41.802281</td>\n",
       "      <td>37.422053</td>\n",
       "      <td>63.293726</td>\n",
       "      <td>PB1</td>\n",
       "      <td>12381</td>\n",
       "      <td>10159</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>5.064007</td>\n",
       "      <td>8277.116060</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000266</td>\n",
       "      <td>0.059091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30899</th>\n",
       "      <td>11530151</td>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>MH8924_B</td>\n",
       "      <td>41.526059</td>\n",
       "      <td>37.491857</td>\n",
       "      <td>54.752443</td>\n",
       "      <td>PB2</td>\n",
       "      <td>2573</td>\n",
       "      <td>10154</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>5.064007</td>\n",
       "      <td>8277.116060</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.051192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31083</th>\n",
       "      <td>333905</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MH8589_A</td>\n",
       "      <td>41.371743</td>\n",
       "      <td>37.513564</td>\n",
       "      <td>56.544991</td>\n",
       "      <td>HA</td>\n",
       "      <td>4006</td>\n",
       "      <td>11508</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>4.250753</td>\n",
       "      <td>1272.405103</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000258</td>\n",
       "      <td>0.131304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31294</th>\n",
       "      <td>9707154</td>\n",
       "      <td>13594</td>\n",
       "      <td>108</td>\n",
       "      <td>MH8586_B</td>\n",
       "      <td>41.739343</td>\n",
       "      <td>37.462415</td>\n",
       "      <td>63.097299</td>\n",
       "      <td>PB1</td>\n",
       "      <td>11617</td>\n",
       "      <td>10980</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>4.193410</td>\n",
       "      <td>1115.017552</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.167685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31471</th>\n",
       "      <td>9481195</td>\n",
       "      <td>1</td>\n",
       "      <td>152</td>\n",
       "      <td>MH8922_A</td>\n",
       "      <td>40.993617</td>\n",
       "      <td>37.063830</td>\n",
       "      <td>65.325532</td>\n",
       "      <td>PB1</td>\n",
       "      <td>8441</td>\n",
       "      <td>11538</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>6.056122</td>\n",
       "      <td>81281.923343</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000300</td>\n",
       "      <td>0.026245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31491</th>\n",
       "      <td>10661195</td>\n",
       "      <td>3</td>\n",
       "      <td>163</td>\n",
       "      <td>MH8922_B</td>\n",
       "      <td>41.704918</td>\n",
       "      <td>37.379098</td>\n",
       "      <td>63.010246</td>\n",
       "      <td>PB1</td>\n",
       "      <td>10422</td>\n",
       "      <td>10449</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>6.056122</td>\n",
       "      <td>81281.923343</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.024628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31607</th>\n",
       "      <td>8520156</td>\n",
       "      <td>13581</td>\n",
       "      <td>70</td>\n",
       "      <td>MH8690_B</td>\n",
       "      <td>41.493766</td>\n",
       "      <td>37.441397</td>\n",
       "      <td>67.558603</td>\n",
       "      <td>PA</td>\n",
       "      <td>7943</td>\n",
       "      <td>4030</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>4.267601</td>\n",
       "      <td>1322.734812</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.121185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31655</th>\n",
       "      <td>10747156</td>\n",
       "      <td>5</td>\n",
       "      <td>104</td>\n",
       "      <td>MH8690_B</td>\n",
       "      <td>41.378205</td>\n",
       "      <td>37.435897</td>\n",
       "      <td>62.705128</td>\n",
       "      <td>PB1</td>\n",
       "      <td>11890</td>\n",
       "      <td>8614</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>4.267601</td>\n",
       "      <td>1322.734812</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000146</td>\n",
       "      <td>0.042952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31793</th>\n",
       "      <td>6904252</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>MH8592_A</td>\n",
       "      <td>40.975066</td>\n",
       "      <td>37.384514</td>\n",
       "      <td>67.385827</td>\n",
       "      <td>NR</td>\n",
       "      <td>2067</td>\n",
       "      <td>905</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>6.547221</td>\n",
       "      <td>251821.707282</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31794</th>\n",
       "      <td>6904255</td>\n",
       "      <td>3</td>\n",
       "      <td>94</td>\n",
       "      <td>MH8592_A</td>\n",
       "      <td>41.585143</td>\n",
       "      <td>37.517714</td>\n",
       "      <td>65.872000</td>\n",
       "      <td>PB1</td>\n",
       "      <td>11330</td>\n",
       "      <td>8832</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>6.547221</td>\n",
       "      <td>251821.707282</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000595</td>\n",
       "      <td>0.031160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31795</th>\n",
       "      <td>6904258</td>\n",
       "      <td>6</td>\n",
       "      <td>120</td>\n",
       "      <td>MH8592_A</td>\n",
       "      <td>30.064327</td>\n",
       "      <td>37.286550</td>\n",
       "      <td>58.122807</td>\n",
       "      <td>PB2</td>\n",
       "      <td>1746</td>\n",
       "      <td>684</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>6.547221</td>\n",
       "      <td>251821.707282</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000412</td>\n",
       "      <td>0.051413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31904</th>\n",
       "      <td>7339752</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>41.546301</td>\n",
       "      <td>37.384365</td>\n",
       "      <td>62.442997</td>\n",
       "      <td>HA</td>\n",
       "      <td>11749</td>\n",
       "      <td>12331</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000291</td>\n",
       "      <td>0.056045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31905</th>\n",
       "      <td>7339754</td>\n",
       "      <td>3</td>\n",
       "      <td>93</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>41.447950</td>\n",
       "      <td>37.233438</td>\n",
       "      <td>63.975815</td>\n",
       "      <td>NR</td>\n",
       "      <td>2009</td>\n",
       "      <td>1512</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.040658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31906</th>\n",
       "      <td>7339755</td>\n",
       "      <td>4</td>\n",
       "      <td>120</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>39.696233</td>\n",
       "      <td>37.439854</td>\n",
       "      <td>57.901580</td>\n",
       "      <td>PA</td>\n",
       "      <td>3255</td>\n",
       "      <td>7470</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000559</td>\n",
       "      <td>0.055406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31907</th>\n",
       "      <td>7339756</td>\n",
       "      <td>5</td>\n",
       "      <td>121</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>39.934831</td>\n",
       "      <td>37.459551</td>\n",
       "      <td>57.384270</td>\n",
       "      <td>PA</td>\n",
       "      <td>3692</td>\n",
       "      <td>7758</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>0.057022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31908</th>\n",
       "      <td>7339757</td>\n",
       "      <td>6</td>\n",
       "      <td>170</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>41.605440</td>\n",
       "      <td>37.343612</td>\n",
       "      <td>61.480751</td>\n",
       "      <td>PB1</td>\n",
       "      <td>11248</td>\n",
       "      <td>9367</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000097</td>\n",
       "      <td>0.278766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31909</th>\n",
       "      <td>7339758</td>\n",
       "      <td>7</td>\n",
       "      <td>189</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>41.413333</td>\n",
       "      <td>37.532549</td>\n",
       "      <td>59.916863</td>\n",
       "      <td>PB2</td>\n",
       "      <td>11711</td>\n",
       "      <td>12809</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>0.000245</td>\n",
       "      <td>0.046524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32099</th>\n",
       "      <td>7353330</td>\n",
       "      <td>13579</td>\n",
       "      <td>38</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>40.918681</td>\n",
       "      <td>37.307114</td>\n",
       "      <td>67.154772</td>\n",
       "      <td>M</td>\n",
       "      <td>14685</td>\n",
       "      <td>3443</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2014-2015</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.449394</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>713 rows  51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0      X  Unnamed..0.1        Id       MapQ      Phred  \\\n",
       "67           4917      1            68    1168.0  41.533913  37.458841   \n",
       "103          7250      2           115    1168.0  41.674238  37.528392   \n",
       "109          7753      4           122    1168.0  41.498099  37.505703   \n",
       "138          9660      5           169    1168.0  41.594608  37.274020   \n",
       "141          9709      6           170    1168.0  41.669893  37.404982   \n",
       "208         12861  13493           241    1168.0  41.618057  37.460768   \n",
       "234         14022      0            13    1121.0  38.713415  37.501220   \n",
       "331         21902  13452           111    1121.0  41.534214  37.391014   \n",
       "479         30573     11            60    1217.0  41.729167  37.629960   \n",
       "493         31635      7            57    1217.0  40.702510  37.346636   \n",
       "500         32073      8            62    1217.0  40.052882  37.469222   \n",
       "571         37668     32           175    1217.0  41.689053  37.503694   \n",
       "588         39151     36           198    1217.0  39.587903  37.444355   \n",
       "713         49574      1           109    1150.0  41.491339  37.338583   \n",
       "715         49578      2           110    1150.0  41.744541  37.207424   \n",
       "857         59799  13448            69    1213.0  36.866667  37.377778   \n",
       "1017        74612      2            79    1125.0  41.608527  37.136628   \n",
       "1066        76725      3           103    1125.0  41.486762  37.289206   \n",
       "1328        94524      0             8    1179.0  39.365166  37.465611   \n",
       "1360        96618      2            40    1179.0  40.964374  37.372825   \n",
       "1476       105503      4           154    1179.0  39.769706  37.250386   \n",
       "1493       106090      5           171    1179.0  41.679124  37.328608   \n",
       "1640       116682      1           108    1154.0  39.007712  37.182519   \n",
       "2206       148508      0             8    1139.0  39.331133  37.403740   \n",
       "2242       151784      1            45    1139.0  41.789398  37.477077   \n",
       "2351       158060      2           147    1139.0  41.801527  37.264122   \n",
       "2490       163867      0            38    1174.0  41.252910  37.413757   \n",
       "2507       164821      2            53    1174.0  41.427141  37.454950   \n",
       "2650       171737      3           200    1174.0  41.652006  37.199846   \n",
       "3157       200805      4           228    1162.0  41.762020  37.545410   \n",
       "...           ...    ...           ...       ...        ...        ...   \n",
       "30352     9489147      7            85  MH8692_A  41.063368  36.944469   \n",
       "30378    12705147     10           113  MH8692_A  41.463320  37.500386   \n",
       "30385    13024147     11           119  MH8692_A  41.003210  37.369181   \n",
       "30452     8229148      0            70  MH8700_A  41.106218  37.446891   \n",
       "30508    10011029  13558             8  MH9045_A  41.616683  37.487918   \n",
       "30576     7594149      2           106  MH9045_A  41.651429  37.492381   \n",
       "30647    12323152      7           229  MH9045_A  41.553222  37.402607   \n",
       "30712     4945150      3            39  MH8925_A  41.588921  37.536443   \n",
       "30725     6241190      4            52  MH8925_A  41.602174  37.639130   \n",
       "30780    11021190      7           122  MH8925_A  40.516971  37.501305   \n",
       "30789    12127150      8           132  MH8925_A  41.527886  37.487678   \n",
       "30855     7280151      0            57  MH8924_B  41.671346  37.517754   \n",
       "30882    10263154      1            90  MH8924_B  41.802281  37.422053   \n",
       "30899    11530151      2           100  MH8924_B  41.526059  37.491857   \n",
       "31083      333905      0             1  MH8589_A  41.371743  37.513564   \n",
       "31294     9707154  13594           108  MH8586_B  41.739343  37.462415   \n",
       "31471     9481195      1           152  MH8922_A  40.993617  37.063830   \n",
       "31491    10661195      3           163  MH8922_B  41.704918  37.379098   \n",
       "31607     8520156  13581            70  MH8690_B  41.493766  37.441397   \n",
       "31655    10747156      5           104  MH8690_B  41.378205  37.435897   \n",
       "31793     6904252      0            41  MH8592_A  40.975066  37.384514   \n",
       "31794     6904255      3            94  MH8592_A  41.585143  37.517714   \n",
       "31795     6904258      6           120  MH8592_A  30.064327  37.286550   \n",
       "31904     7339752      1            19  MH8766_B  41.546301  37.384365   \n",
       "31905     7339754      3            93  MH8766_B  41.447950  37.233438   \n",
       "31906     7339755      4           120  MH8766_B  39.696233  37.439854   \n",
       "31907     7339756      5           121  MH8766_B  39.934831  37.459551   \n",
       "31908     7339757      6           170  MH8766_B  41.605440  37.343612   \n",
       "31909     7339758      7           189  MH8766_B  41.413333  37.532549   \n",
       "32099     7353330  13579            38  MH8766_B  40.918681  37.307114   \n",
       "\n",
       "        Read_pos  chr  cov.ctrl.bw  cov.ctrl.fw    ...     DPI     season  \\\n",
       "67     61.382029   NR        18013        18212    ...       1      10-11   \n",
       "103    59.034668   PA        15825        17751    ...       1      10-11   \n",
       "109    62.321466   PA        18084        17365    ...       1      10-11   \n",
       "138    63.987255  PB1        27047        23837    ...       1      10-11   \n",
       "141    62.524217  PB1        25999        27035    ...       1      10-11   \n",
       "208    64.205162  PB2        25826        24738    ...       1      10-11   \n",
       "234    62.269512   HA        17196        19477    ...       2      10-11   \n",
       "331    65.533185   PA        18262        11558    ...       2      10-11   \n",
       "479    64.272817   NP        30202        29646    ...       1      11-12   \n",
       "493    59.452422   NR         8118        12397    ...       1      11-12   \n",
       "500    61.655568   NR        12759        12972    ...       1      11-12   \n",
       "571    64.870719  PB1        26113        19314    ...       1      11-12   \n",
       "588    60.845161  PB2        23892        26738    ...       1      11-12   \n",
       "713    58.414173  PB1        10730        19197    ...       2      10-11   \n",
       "715    57.943231  PB1        10774        18512    ...       2      10-11   \n",
       "857    59.311111   NS         7369        24610    ...       0      11-12   \n",
       "1017   61.093346   PA        10743        11016    ...       3      10-11   \n",
       "1066   63.089613  PB1        18948        19386    ...       3      10-11   \n",
       "1328   56.783276   HA         3281        10176    ...       0      10-11   \n",
       "1360   63.579950    M        15210        12070    ...       0      10-11   \n",
       "1476   32.769706  PB2          498         2409    ...       0      10-11   \n",
       "1493   63.149485  PB2        18448        18166    ...       0      10-11   \n",
       "1640   33.053985  PB1         1195         5654    ...       2      10-11   \n",
       "2206   57.577558   HA         5812        12753    ...       2      10-11   \n",
       "2242   60.324737   NP        21150        22814    ...       2      10-11   \n",
       "2351   63.155725  PB1        19342        19402    ...       2      10-11   \n",
       "2490   62.187302    M        14105        15148    ...       1      10-11   \n",
       "2507   59.296997   NP        13767        21388    ...       1      10-11   \n",
       "2650   60.466821  PB1        18770        20262    ...       1      10-11   \n",
       "3157   61.382224  PB2        25812        25609    ...       1      10-11   \n",
       "...          ...  ...          ...          ...    ...     ...        ...   \n",
       "30352  60.461039  PB1        10798        11046    ...       4  2014-2015   \n",
       "30378  63.997683  PB2        11916        13281    ...       4  2014-2015   \n",
       "30385  63.711075  PB2        12266        10707    ...       4  2014-2015   \n",
       "30452  63.032383   PA         8186         7839    ...       4  2014-2015   \n",
       "30508  63.541377   HA        11606        12259    ...       2  2014-2015   \n",
       "30576  62.947619   PA         8430         7612    ...       2  2014-2015   \n",
       "30647  60.401159  PB2        10801        12104    ...       2  2014-2015   \n",
       "30712  61.093294   NR         1914         1892    ...       5  2014-2015   \n",
       "30725  61.995652   NS        18029        17069    ...       5  2014-2015   \n",
       "30780  75.872063  PB1         6384         1206    ...       5  2014-2015   \n",
       "30789  61.630350  PB2        11121        11413    ...       5  2014-2015   \n",
       "30855  64.987614   PA         7607         7617    ...       4  2014-2015   \n",
       "30882  63.293726  PB1        12381        10159    ...       4  2014-2015   \n",
       "30899  54.752443  PB2         2573        10154    ...       4  2014-2015   \n",
       "31083  56.544991   HA         4006        11508    ...       5  2014-2015   \n",
       "31294  63.097299  PB1        11617        10980    ...       3  2014-2015   \n",
       "31471  65.325532  PB1         8441        11538    ...       1  2014-2015   \n",
       "31491  63.010246  PB1        10422        10449    ...       1  2014-2015   \n",
       "31607  67.558603   PA         7943         4030    ...       6  2014-2015   \n",
       "31655  62.705128  PB1        11890         8614    ...       6  2014-2015   \n",
       "31793  67.385827   NR         2067          905    ...       2  2014-2015   \n",
       "31794  65.872000  PB1        11330         8832    ...       2  2014-2015   \n",
       "31795  58.122807  PB2         1746          684    ...       2  2014-2015   \n",
       "31904  62.442997   HA        11749        12331    ...       4  2014-2015   \n",
       "31905  63.975815   NR         2009         1512    ...       4  2014-2015   \n",
       "31906  57.901580   PA         3255         7470    ...       4  2014-2015   \n",
       "31907  57.384270   PA         3692         7758    ...       4  2014-2015   \n",
       "31908  61.480751  PB1        11248         9367    ...       4  2014-2015   \n",
       "31909  59.916863  PB2        11711        12809    ...       4  2014-2015   \n",
       "32099  67.154772    M        14685         3443    ...       4  2014-2015   \n",
       "\n",
       "       log_copy_num          gc_ul  HIGHSD  sequenced  home_collected  \\\n",
       "67         6.136662   97843.843292       N       True               0   \n",
       "103        6.136662   97843.843292       N       True               0   \n",
       "109        6.136662   97843.843292       N       True               0   \n",
       "138        6.136662   97843.843292       N       True               0   \n",
       "141        6.136662   97843.843292       N       True               0   \n",
       "208        6.136662   97843.843292       N       True               0   \n",
       "234        5.060653    8213.445551       N       True               0   \n",
       "331        5.060653    8213.445551       N       True               0   \n",
       "479        4.240458    1242.597283       N       True               0   \n",
       "493        4.240458    1242.597283       N       True               0   \n",
       "500        4.240458    1242.597283       N       True               0   \n",
       "571        4.240458    1242.597283       N       True               0   \n",
       "588        4.240458    1242.597283       N       True               0   \n",
       "713        5.922566   59763.735122       N       True               0   \n",
       "715        5.922566   59763.735122       N       True               0   \n",
       "857        5.542293   24898.041190       N       True               0   \n",
       "1017       4.851537    5074.685473       N       True               0   \n",
       "1066       4.851537    5074.685473       N       True               0   \n",
       "1328       5.621547   29882.647245       N       True               0   \n",
       "1360       5.621547   29882.647245       N       True               0   \n",
       "1476       5.621547   29882.647245       N       True               0   \n",
       "1493       5.621547   29882.647245       N       True               0   \n",
       "1640       4.931225    6096.725833       N       True               0   \n",
       "2206       5.288065   13865.536380       N       True               0   \n",
       "2242       5.288065   13865.536380       N       True               0   \n",
       "2351       5.288065   13865.536380       N       True               0   \n",
       "2490       5.318185   14861.296029       N       True               0   \n",
       "2507       5.318185   14861.296029       N       True               0   \n",
       "2650       5.318185   14861.296029       N       True               0   \n",
       "3157       4.812868    4642.371815       N       True               0   \n",
       "...             ...            ...     ...        ...             ...   \n",
       "30352      4.418474    1872.171013       N       True               0   \n",
       "30378      4.418474    1872.171013       N       True               0   \n",
       "30385      4.418474    1872.171013       N       True               0   \n",
       "30452      5.462005   20695.554297       N       True               0   \n",
       "30508      5.285284   13777.027742       N       True               0   \n",
       "30576      5.285284   13777.027742       N       True               0   \n",
       "30647      5.285284   13777.027742       N       True               0   \n",
       "30712      5.684998   34583.562088       N       True               0   \n",
       "30725      5.684998   34583.562088       N       True               0   \n",
       "30780      5.684998   34583.562088       N       True               0   \n",
       "30789      5.684998   34583.562088       N       True               0   \n",
       "30855      5.064007    8277.116060       N       True               0   \n",
       "30882      5.064007    8277.116060       N       True               0   \n",
       "30899      5.064007    8277.116060       N       True               0   \n",
       "31083      4.250753    1272.405103       N       True               0   \n",
       "31294      4.193410    1115.017552       N       True               0   \n",
       "31471      6.056122   81281.923343       N       True               0   \n",
       "31491      6.056122   81281.923343       N       True               0   \n",
       "31607      4.267601    1322.734812       N       True               0   \n",
       "31655      4.267601    1322.734812       N       True               0   \n",
       "31793      6.547221  251821.707282       N       True               0   \n",
       "31794      6.547221  251821.707282       N       True               0   \n",
       "31795      6.547221  251821.707282       N       True               0   \n",
       "31904      6.371922  168187.646394       N       True               0   \n",
       "31905      6.371922  168187.646394       N       True               0   \n",
       "31906      6.371922  168187.646394       N       True               0   \n",
       "31907      6.371922  168187.646394       N       True               0   \n",
       "31908      6.371922  168187.646394       N       True               0   \n",
       "31909      6.371922  168187.646394       N       True               0   \n",
       "32099      6.371922  168187.646394       N       True               0   \n",
       "\n",
       "        class_factor  control.freq  exp.freq  \n",
       "67        Synonymous      0.000193  0.063075  \n",
       "103    Nonsynonymous      0.000060  0.045823  \n",
       "109       Synonymous      0.000056  0.067699  \n",
       "138       Synonymous      0.000157  0.045392  \n",
       "141    Nonsynonymous      0.000170  0.462501  \n",
       "208       Synonymous           NaN  0.445909  \n",
       "234       Synonymous      0.000109  0.020245  \n",
       "331       Synonymous           NaN  0.260802  \n",
       "479    Nonsynonymous      0.000184  0.026508  \n",
       "493       Synonymous      0.000146  0.269504  \n",
       "500    Nonsynonymous      0.000272  0.137748  \n",
       "571       Synonymous      0.000286  0.131039  \n",
       "588       Synonymous      0.000138  0.045310  \n",
       "713    Nonsynonymous      0.000100  0.035542  \n",
       "715       Synonymous      0.000137  0.026382  \n",
       "857    Nonsynonymous           NaN  0.022819  \n",
       "1017   Nonsynonymous      0.000322  0.210355  \n",
       "1066      Synonymous      0.000157  0.028421  \n",
       "1328   Nonsynonymous      0.000297  0.071512  \n",
       "1360   Nonsynonymous      0.000513  0.028234  \n",
       "1476   Nonsynonymous      0.000688  0.214451  \n",
       "1493   Nonsynonymous      0.000137  0.021868  \n",
       "1640      Synonymous      0.000292  0.196564  \n",
       "2206      Synonymous      0.000162  0.032483  \n",
       "2242      Synonymous      0.000205  0.056103  \n",
       "2351      Synonymous      0.000310  0.024857  \n",
       "2490   Nonsynonymous      0.000308  0.021896  \n",
       "2507   Nonsynonymous      0.000398  0.025191  \n",
       "2650      Synonymous      0.000051  0.038458  \n",
       "3157   Nonsynonymous      0.000136  0.047676  \n",
       "...              ...           ...       ...  \n",
       "30352     Synonymous      0.000229  0.174161  \n",
       "30378  Nonsynonymous      0.000317  0.041392  \n",
       "30385     Synonymous      0.000131  0.022652  \n",
       "30452     Synonymous      0.000187  0.024680  \n",
       "30508  Nonsynonymous           NaN  0.160589  \n",
       "30576  Nonsynonymous      0.000187  0.036735  \n",
       "30647  Nonsynonymous      0.000131  0.059973  \n",
       "30712     Synonymous      0.000000  0.021167  \n",
       "30725  Nonsynonymous      0.000342  0.027553  \n",
       "30780     Synonymous      0.000791  0.044519  \n",
       "30789  Nonsynonymous      0.000089  0.029041  \n",
       "30855  Nonsynonymous      0.000263  0.057796  \n",
       "30882     Synonymous      0.000266  0.059091  \n",
       "30899  Nonsynonymous      0.000000  0.051192  \n",
       "31083     Synonymous      0.000258  0.131304  \n",
       "31294     Synonymous           NaN  0.167685  \n",
       "31471     Synonymous      0.000300  0.026245  \n",
       "31491     Synonymous      0.000048  0.024628  \n",
       "31607     Synonymous           NaN  0.121185  \n",
       "31655     Synonymous      0.000146  0.042952  \n",
       "31793     Synonymous      0.000000  0.030573  \n",
       "31794     Synonymous      0.000595  0.031160  \n",
       "31795  Nonsynonymous      0.000412  0.051413  \n",
       "31904     Synonymous      0.000291  0.056045  \n",
       "31905  Nonsynonymous      0.000000  0.040658  \n",
       "31906     Synonymous      0.000559  0.055406  \n",
       "31907     Synonymous      0.000175  0.057022  \n",
       "31908     Synonymous      0.000097  0.278766  \n",
       "31909     Synonymous      0.000245  0.046524  \n",
       "32099  Nonsynonymous           NaN  0.449394  \n",
       "\n",
       "[713 rows x 51 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a function that locates each HA mutation in the PDB structures. This is the equivalent of H3 numbering for H3 strains. (Assigns codon 1 to codon 17 or (AA pos - 16))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def locate(snv,run_label,ref_fa,antigenic):\n",
    "    # Get the HA seqeunce the sample was aligned to - the cordinates here match those in the dataframe\n",
    "    for seq_record in SeqIO.parse(ref_fa, \"fasta\"):    \n",
    "        if seq_record.name == \"HA\":\n",
    "            HA=seq_record\n",
    "            HA_prot = seq_record.seq.translate(to_stop=True) # translate to protien.\n",
    "    run_snv = snv.loc[(snv[\"run\"].str.contains(run_label))  & (snv[\"chr\"]==\"HA\")] # Get the isnv from that run and that in HA \n",
    "    PDB_4HMG=[]\n",
    "    PDB_4JTV=[]\n",
    "    for index, var in run_snv.iterrows():\n",
    "        nice = find_aligments('muscle',\"/Users/jt/muscle3.8.31/\",HA_prot,var[\"AA_pos\"]) # These positions are base 1 (see below) whcih is what the function expects\n",
    "        PDB_4HMG.append(nice[0][0])\n",
    "        PDB_4JTV.append(nice[1][0])\n",
    "    run_snv[\"PDB_4HMG\"] = PDB_4HMG\n",
    "    run_snv[\"PDB_4JTV\"] = PDB_4JTV\n",
    "    return(run_snv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perth 2010-2011 & 2011-2012"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue H172 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K156 in HA1 in 4HMG\n",
      "  * K159 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E78 in sequential numbering of your HA sequence corresponds to:\n",
      "  * I62 in HA1 in 4HMG\n",
      "  * G59 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue D101 in sequential numbering of your HA sequence corresponds to:\n",
      "  * D85 in HA1 in 4HMG\n",
      "  * S83 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue V313 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V297 in HA1 in 4HMG\n",
      "  * I301 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E105 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E89 in HA1 in 4HMG\n",
      "  * E87 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L102 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L86 in HA1 in 4HMG\n",
      "  * Y84 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A554 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I10 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue N24 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N8 in HA1 in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K557 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "perth = locate(snv,\"perth\",\"/Users/jt/Documents/Analysis/HIVE/data/processed/perth/coding_fa/Perth_mp.removed.parsed.coding.fasta\",antigenic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Victoria 2012-2013 & 2013-2014"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I230 in sequential numbering of your HA sequence corresponds to:\n",
      "  * I214 in HA1 in 4HMG\n",
      "  * K217 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T133 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T117 in HA1 in 4HMG\n",
      "  * R119 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K524 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue G197 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G181 in HA1 in 4HMG\n",
      "  * G184 of HA1 in 4JTV\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "victoria = locate(snv,\"vic\",\"/Users/jt/Documents/Analysis/HIVE/data/processed/victoria/coding_fa/Vic_pool.removed.parsed.coding.fasta\",antigenic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hong Kong 2014-2015"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T40 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T24 in HA1 in 4HMG\n",
      "  * D20 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue G302 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G286 in HA1 in 4HMG\n",
      "  * G290 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R315 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K299 in HA1 in 4HMG\n",
      "  * P303 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue V429 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V84 in HA2 in 4HMG\n",
      "  * V84 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue N499 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N154 in HA2 in 4HMG\n",
      "  * N154 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue F209 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S193 in HA1 in 4HMG\n",
      "  * S196 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue H172 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K156 in HA1 in 4HMG\n",
      "  * K159 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T144 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T128 in HA1 in 4HMG\n",
      "  * P130 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S165 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S149 in HA1 in 4HMG\n",
      "  * K152 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T203 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T187 in HA1 in 4HMG\n",
      "  * T190 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T40 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T24 in HA1 in 4HMG\n",
      "  * D20 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E188 in sequential numbering of your HA sequence corresponds to:\n",
      "  * D172 in HA1 in 4HMG\n",
      "  * K175 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I218 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V202 in HA1 in 4HMG\n",
      "  * V205 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I276 in sequential numbering of your HA sequence corresponds to:\n",
      "  * M260 in HA1 in 4HMG\n",
      "  * M263 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E296 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E280 in HA1 in 4HMG\n",
      "  * T284 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue G302 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G286 in HA1 in 4HMG\n",
      "  * G290 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R315 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K299 in HA1 in 4HMG\n",
      "  * P303 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue V429 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V84 in HA2 in 4HMG\n",
      "  * V84 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue V339 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V323 in HA1 in 4HMG\n",
      "  * I327 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S278 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T262 in HA1 in 4HMG\n",
      "  * N266 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K468 in sequential numbering of your HA sequence corresponds to:\n",
      "  * R123 in HA2 in 4HMG\n",
      "  * R123 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L383 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L38 in HA2 in 4HMG\n",
      "  * L38 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue C482 in sequential numbering of your HA sequence corresponds to:\n",
      "  * C137 in HA2 in 4HMG\n",
      "  * C137 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue F415 in sequential numbering of your HA sequence corresponds to:\n",
      "  * F70 in HA2 in 4HMG\n",
      "  * F70 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S165 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S149 in HA1 in 4HMG\n",
      "  * K152 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T203 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T187 in HA1 in 4HMG\n",
      "  * T190 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L330 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L314 in HA1 in 4HMG\n",
      "  * L318 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P37 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P21 in HA1 in 4HMG\n",
      "  * N17 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P322 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P306 in HA1 in 4HMG\n",
      "  * P310 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L547 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S130 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S114 in HA1 in 4HMG\n",
      "  * S116 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A154 in sequential numbering of your HA sequence corresponds to:\n",
      "  * A138 in HA1 in 4HMG\n",
      "  * A141 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A212 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V196 in HA1 in 4HMG\n",
      "  * Q199 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E465 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E120 in HA2 in 4HMG\n",
      "  * E120 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A6 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A16 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K108 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K92 in HA1 in 4HMG\n",
      "  * S91 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E188 in sequential numbering of your HA sequence corresponds to:\n",
      "  * D172 in HA1 in 4HMG\n",
      "  * K175 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I218 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V202 in HA1 in 4HMG\n",
      "  * V205 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I218 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V202 in HA1 in 4HMG\n",
      "  * V205 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R224 in sequential numbering of your HA sequence corresponds to:\n",
      "  * R208 in HA1 in 4HMG\n",
      "  * R211 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E296 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E280 in HA1 in 4HMG\n",
      "  * T284 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A179 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V163 in HA1 in 4HMG\n",
      "  * K166 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L332 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L316 in HA1 in 4HMG\n",
      "  * L320 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P340 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P324 in HA1 in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue G495 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E150 in HA2 in 4HMG\n",
      "  * E150 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R323 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K307 in HA1 in 4HMG\n",
      "  * K311 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue M478 in sequential numbering of your HA sequence corresponds to:\n",
      "  * M133 in HA2 in 4HMG\n",
      "  * I133 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P322 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P306 in HA1 in 4HMG\n",
      "  * P310 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue H451 in sequential numbering of your HA sequence corresponds to:\n",
      "  * H106 in HA2 in 4HMG\n",
      "  * R106 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K205 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Q189 in HA1 in 4HMG\n",
      "  * A192 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue H172 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K156 in HA1 in 4HMG\n",
      "  * K159 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L82 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L66 in HA1 in 4HMG\n",
      "  * I63 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue F348 in sequential numbering of your HA sequence corresponds to:\n",
      "  * F3 in HA2 in 4HMG\n",
      "  * F3 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue M284 in sequential numbering of your HA sequence corresponds to:\n",
      "  * M268 in HA1 in 4HMG\n",
      "  * I272 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue D69 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N53 in HA1 in 4HMG\n",
      "  * K49 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S328 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N312 in HA1 in 4HMG\n",
      "  * T316 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue Y486 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Y141 in HA2 in 4HMG\n",
      "  * Y141 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I258 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V242 in HA1 in 4HMG\n",
      "  * K245 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A122 in sequential numbering of your HA sequence corresponds to:\n",
      "  * A106 in HA1 in 4HMG\n",
      "  * E105 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I258 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V242 in HA1 in 4HMG\n",
      "  * K245 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P37 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P21 in HA1 in 4HMG\n",
      "  * N17 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L547 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue D307 in sequential numbering of your HA sequence corresponds to:\n",
      "  * D291 in HA1 in 4HMG\n",
      "  * S295 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A288 in sequential numbering of your HA sequence corresponds to:\n",
      "  * A272 in HA1 in 4HMG\n",
      "  * T276 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R323 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K307 in HA1 in 4HMG\n",
      "  * K311 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue D101 in sequential numbering of your HA sequence corresponds to:\n",
      "  * D85 in HA1 in 4HMG\n",
      "  * S83 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue G279 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G263 in HA1 in 4HMG\n",
      "  * A267 of HA1 in 4JTV\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "HK = locate(snv,\"HK\",\"/Users/jt/Documents/Analysis/HIVE/data/processed/HK_6/coding_fa/PC1A.removed.parsed.coding.fasta\",antigenic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cali09 H1N1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue Q210 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V196 in HA1 in 4HMG\n",
      "  * Q199 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S145 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Q132 in HA1 in 4HMG\n",
      "  * S134 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T451 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T107 in HA2 in 4HMG\n",
      "  * T107 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue G172 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G158 in HA1 in 4HMG\n",
      "  * G161 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue F432 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K88 in HA2 in 4HMG\n",
      "  * F88 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue D489 in sequential numbering of your HA sequence corresponds to:\n",
      "  * D145 in HA2 in 4HMG\n",
      "  * D145 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue V149 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * V138 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K511 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L167 in HA2 in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R130 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T117 in HA1 in 4HMG\n",
      "  * R119 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E241 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S227 in HA1 in 4HMG\n",
      "  * E230 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T262 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N248 in HA1 in 4HMG\n",
      "  * T251 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L442 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L98 in HA2 in 4HMG\n",
      "  * L98 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R222 in sequential numbering of your HA sequence corresponds to:\n",
      "  * R208 in HA1 in 4HMG\n",
      "  * R211 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue N404 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N60 in HA2 in 4HMG\n",
      "  * N60 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue G23 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G16 in HA1 in 4HMG\n",
      "  * G12 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E132 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E119 in HA1 in 4HMG\n",
      "  * E121 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P199 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P185 in HA1 in 4HMG\n",
      "  * P188 of HA1 in 4JTV\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "cali09 = locate(snv,\"cali\",\"/Users/jt/Documents/Analysis/HIVE/data/processed/cali09/coding_fa/Cali_pool.removed.parsed.coding.fasta\",antigenic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# All sites\n",
    "\n",
    "Here we combine all the runs together, select for the nonsynonymous mutations and make a subtype column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  app.launch_new_instance()\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/pandas/core/indexing.py:465: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self.obj[item] = s\n"
     ]
    }
   ],
   "source": [
    "HA_All = perth.append([victoria,HK,cali09], ignore_index = True)\n",
    "HA_nonsyn = HA_All.loc[HA_All[\"class_factor\"]==\"Nonsynonymous\"]\n",
    "HA_nonsyn[\"subtype\"] = \"H3\"\n",
    "HA_nonsyn.loc[HA_nonsyn[\"run\"]==\"cali09\",\"subtype\"]=\"H1\"\n",
    "HA_nonsyn.loc[HA_nonsyn[\"run\"]==\"cali09_2\",\"subtype\"]=\"H1\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will us regex to split the PDB names into amino acid position and segement (HA1- HA2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "loc=[]\n",
    "seg = []\n",
    "pat = re.compile('\\w(\\d+) in (HA[1,2])') \n",
    "for index, row in HA_nonsyn.iterrows():\n",
    "    s = pat.match(row.loc[\"PDB_4HMG\"])\n",
    "    if s:\n",
    "        loc.append(s.group(1))\n",
    "        seg.append(s.group(2))\n",
    "    else:\n",
    "        loc.append(\"-\")\n",
    "        seg.append(\"-\")\n",
    "HA_nonsyn[\"H3_pos\"]=loc\n",
    "HA_nonsyn[\"H3_seg\"]=seg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, which of these positions are in antigenic sites?\n",
    "\n",
    "For this work I am cycling through each row of the data frame. \n",
    "    \n",
    "If the row is for an H3 virus in with a mutation in the HA1 segment Then we find the antigenic site in the frist try statement. I then assign the source in the second try statement.\n",
    "    \n",
    "    \n",
    "If the row is in an H1 virus then we check if the PBD_4JTV position is in the PBD_4JTV list of antigenic sites. All these came from Caton."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:28: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "antigenic_sites=[]\n",
    "source = []\n",
    "for index, row in HA_nonsyn.iterrows():\n",
    "    if row.loc[\"subtype\"]==\"H3\" and row.loc[\"H3_seg\"]==\"HA1\":\n",
    "        try:\n",
    "            antigenic_sites.append(H3.loc[H3['aa']==int(row.loc[\"H3_pos\"]),\"Antigenic sites\"].values[0])\n",
    "        except IndexError:\n",
    "            antigenic_sites.append(\"NA\")\n",
    "        try:\n",
    "            s = H3.loc[H3['aa']==int(row.loc[\"H3_pos\"]),\"Source\"].values[0]\n",
    "            if pd.isnull(s): # If there is no input then the site was mentioned in Lee et al. 2004\n",
    "                s=\"Lee\"\n",
    "            source.append(s)\n",
    "        except IndexError:\n",
    "            source.append(\"NA\")\n",
    "        \n",
    "    elif row.loc[\"subtype\"]==\"H1\":\n",
    "        if row.loc[\"PDB_4JTV\"] in antigenic:\n",
    "            antigenic_sites.append(\"H1N1\")\n",
    "            source.append(\"Caton\")\n",
    "        else:\n",
    "            antigenic_sites.append(\"NA\")\n",
    "            source.append(\"NA\")\n",
    "    else:\n",
    "        antigenic_sites.append(\"NA\")\n",
    "        source.append(\"NA\")\n",
    "HA_nonsyn[\"Antigenic\"] = antigenic_sites\n",
    "HA_nonsyn[\"Source\"] = source"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here are the nonsynonymous mutations in HA with the antigenic information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SPECID</th>\n",
       "      <th>onset</th>\n",
       "      <th>mutation</th>\n",
       "      <th>freq.var</th>\n",
       "      <th>Ref_AA</th>\n",
       "      <th>AA_pos</th>\n",
       "      <th>Var_AA</th>\n",
       "      <th>PDB_4JTV</th>\n",
       "      <th>PDB_4HMG</th>\n",
       "      <th>run</th>\n",
       "      <th>Antigenic</th>\n",
       "      <th>Source</th>\n",
       "      <th>subtype</th>\n",
       "      <th>H3_pos</th>\n",
       "      <th>H3_seg</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M54913</td>\n",
       "      <td>2011-03-30</td>\n",
       "      <td>HA_A258G</td>\n",
       "      <td>0.071215</td>\n",
       "      <td>['E']</td>\n",
       "      <td>[78]</td>\n",
       "      <td>['G']</td>\n",
       "      <td>G59 of HA1</td>\n",
       "      <td>I62 in HA1</td>\n",
       "      <td>perth_2</td>\n",
       "      <td>E</td>\n",
       "      <td>Smith</td>\n",
       "      <td>H3</td>\n",
       "      <td>62</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MH0778</td>\n",
       "      <td>2012-02-13</td>\n",
       "      <td>HA_T963C</td>\n",
       "      <td>0.071359</td>\n",
       "      <td>['V']</td>\n",
       "      <td>[313]</td>\n",
       "      <td>['A']</td>\n",
       "      <td>I301 of HA1</td>\n",
       "      <td>V297 in HA1</td>\n",
       "      <td>perth_2</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>297</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>M54724</td>\n",
       "      <td>2011-03-20</td>\n",
       "      <td>HA_C329A</td>\n",
       "      <td>0.087608</td>\n",
       "      <td>['L']</td>\n",
       "      <td>[102]</td>\n",
       "      <td>['I']</td>\n",
       "      <td>Y84 of HA1</td>\n",
       "      <td>L86 in HA1</td>\n",
       "      <td>perth</td>\n",
       "      <td>E</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>86</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>M55001</td>\n",
       "      <td>2011-04-01</td>\n",
       "      <td>HA_T54T</td>\n",
       "      <td>0.032427</td>\n",
       "      <td>['T']</td>\n",
       "      <td>[10]</td>\n",
       "      <td>['I']</td>\n",
       "      <td>an alignment gap</td>\n",
       "      <td>an alignment gap</td>\n",
       "      <td>perth</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MH2436</td>\n",
       "      <td>2012-12-10</td>\n",
       "      <td>HA_T718C</td>\n",
       "      <td>0.065651</td>\n",
       "      <td>['I']</td>\n",
       "      <td>[230]</td>\n",
       "      <td>['T']</td>\n",
       "      <td>K217 of HA1</td>\n",
       "      <td>I214 in HA1</td>\n",
       "      <td>vic</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>214</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MH4091</td>\n",
       "      <td>2013-03-05</td>\n",
       "      <td>HA_A1599G</td>\n",
       "      <td>0.079685</td>\n",
       "      <td>['K']</td>\n",
       "      <td>[524]</td>\n",
       "      <td>['E']</td>\n",
       "      <td>an alignment gap</td>\n",
       "      <td>an alignment gap</td>\n",
       "      <td>vic</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MH8308</td>\n",
       "      <td>2014-12-16</td>\n",
       "      <td>HA_T1526A</td>\n",
       "      <td>0.041457</td>\n",
       "      <td>['N']</td>\n",
       "      <td>[499]</td>\n",
       "      <td>['K']</td>\n",
       "      <td>N154 of HA2</td>\n",
       "      <td>N154 in HA2</td>\n",
       "      <td>HK_2</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>154</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MH7851</td>\n",
       "      <td>2014-12-05</td>\n",
       "      <td>HA_T655C</td>\n",
       "      <td>0.174563</td>\n",
       "      <td>['F']</td>\n",
       "      <td>[209]</td>\n",
       "      <td>['S']</td>\n",
       "      <td>S196 of HA1</td>\n",
       "      <td>S193 in HA1</td>\n",
       "      <td>HK_2</td>\n",
       "      <td>B</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>H3</td>\n",
       "      <td>193</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MH7860</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>HA_A459G</td>\n",
       "      <td>0.110522</td>\n",
       "      <td>['T']</td>\n",
       "      <td>[144]</td>\n",
       "      <td>['A']</td>\n",
       "      <td>P130 of HA1</td>\n",
       "      <td>T128 in HA1</td>\n",
       "      <td>HK_8</td>\n",
       "      <td>B</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>128</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>MH8137</td>\n",
       "      <td>2014-12-15</td>\n",
       "      <td>HA_A855G</td>\n",
       "      <td>0.404932</td>\n",
       "      <td>['I']</td>\n",
       "      <td>[276]</td>\n",
       "      <td>['V']</td>\n",
       "      <td>M263 of HA1</td>\n",
       "      <td>M260 in HA1</td>\n",
       "      <td>HK_2</td>\n",
       "      <td>E</td>\n",
       "      <td>Smith</td>\n",
       "      <td>H3</td>\n",
       "      <td>260</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>MH8052</td>\n",
       "      <td>2014-12-13</td>\n",
       "      <td>HA_G862A</td>\n",
       "      <td>0.030123</td>\n",
       "      <td>['S']</td>\n",
       "      <td>[278]</td>\n",
       "      <td>['N']</td>\n",
       "      <td>N266 of HA1</td>\n",
       "      <td>T262 in HA1</td>\n",
       "      <td>HK_8</td>\n",
       "      <td>E</td>\n",
       "      <td>Smith</td>\n",
       "      <td>H3</td>\n",
       "      <td>262</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>HS1595</td>\n",
       "      <td>2015-01-04</td>\n",
       "      <td>HA_C418T</td>\n",
       "      <td>0.031555</td>\n",
       "      <td>['S']</td>\n",
       "      <td>[130]</td>\n",
       "      <td>['L']</td>\n",
       "      <td>S116 of HA1</td>\n",
       "      <td>S114 in HA1</td>\n",
       "      <td>HK_7</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>114</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>HS1530</td>\n",
       "      <td>2014-12-22</td>\n",
       "      <td>HA_G75A</td>\n",
       "      <td>0.369782</td>\n",
       "      <td>['A']</td>\n",
       "      <td>[16]</td>\n",
       "      <td>['T']</td>\n",
       "      <td>an alignment gap</td>\n",
       "      <td>an alignment gap</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>HS1530</td>\n",
       "      <td>2014-12-22</td>\n",
       "      <td>HA_A681G</td>\n",
       "      <td>0.360914</td>\n",
       "      <td>['I']</td>\n",
       "      <td>[218]</td>\n",
       "      <td>['V']</td>\n",
       "      <td>V205 of HA1</td>\n",
       "      <td>V202 in HA1</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>202</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>HS1530</td>\n",
       "      <td>2014-12-22</td>\n",
       "      <td>HA_A699A</td>\n",
       "      <td>0.363668</td>\n",
       "      <td>['G']</td>\n",
       "      <td>[224]</td>\n",
       "      <td>['R']</td>\n",
       "      <td>R211 of HA1</td>\n",
       "      <td>R208 in HA1</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>208</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>MH7755</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>HA_G564A</td>\n",
       "      <td>0.032187</td>\n",
       "      <td>['A']</td>\n",
       "      <td>[179]</td>\n",
       "      <td>['T']</td>\n",
       "      <td>K166 of HA1</td>\n",
       "      <td>V163 in HA1</td>\n",
       "      <td>HK_7</td>\n",
       "      <td>B</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>163</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>MH7796</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>HA_G1512A</td>\n",
       "      <td>0.111589</td>\n",
       "      <td>['G']</td>\n",
       "      <td>[495]</td>\n",
       "      <td>['R']</td>\n",
       "      <td>E150 of HA2</td>\n",
       "      <td>E150 in HA2</td>\n",
       "      <td>HK_7</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>150</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>HS1604</td>\n",
       "      <td>2015-01-11</td>\n",
       "      <td>HA_G997G</td>\n",
       "      <td>0.118966</td>\n",
       "      <td>['I']</td>\n",
       "      <td>[323]</td>\n",
       "      <td>['R']</td>\n",
       "      <td>K311 of HA1</td>\n",
       "      <td>K307 in HA1</td>\n",
       "      <td>HK_7</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>307</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>MH7611</td>\n",
       "      <td>2014-11-23</td>\n",
       "      <td>HA_G1463A</td>\n",
       "      <td>0.043976</td>\n",
       "      <td>['M']</td>\n",
       "      <td>[478]</td>\n",
       "      <td>['I']</td>\n",
       "      <td>I133 of HA2</td>\n",
       "      <td>M133 in HA2</td>\n",
       "      <td>HK_7</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>133</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>HS1455</td>\n",
       "      <td>2014-12-13</td>\n",
       "      <td>HA_A1381A</td>\n",
       "      <td>0.350988</td>\n",
       "      <td>['R']</td>\n",
       "      <td>[451]</td>\n",
       "      <td>['H']</td>\n",
       "      <td>R106 of HA2</td>\n",
       "      <td>H106 in HA2</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>106</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>HS1450</td>\n",
       "      <td>2014-12-13</td>\n",
       "      <td>HA_G644T</td>\n",
       "      <td>0.037827</td>\n",
       "      <td>['K']</td>\n",
       "      <td>[205]</td>\n",
       "      <td>['N']</td>\n",
       "      <td>A192 of HA1</td>\n",
       "      <td>Q189 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>B</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>H3</td>\n",
       "      <td>189</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>HS1473</td>\n",
       "      <td>2014-12-15</td>\n",
       "      <td>HA_T1071C</td>\n",
       "      <td>0.177116</td>\n",
       "      <td>['F']</td>\n",
       "      <td>[348]</td>\n",
       "      <td>['L']</td>\n",
       "      <td>F3 of HA2</td>\n",
       "      <td>F3 in HA2</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>3</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>HS1401</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>HA_A879C</td>\n",
       "      <td>0.034116</td>\n",
       "      <td>['M']</td>\n",
       "      <td>[284]</td>\n",
       "      <td>['L']</td>\n",
       "      <td>I272 of HA1</td>\n",
       "      <td>M268 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>268</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>HS1376</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>HA_C236A</td>\n",
       "      <td>0.025191</td>\n",
       "      <td>['D']</td>\n",
       "      <td>[69]</td>\n",
       "      <td>['E']</td>\n",
       "      <td>K49 of HA1</td>\n",
       "      <td>N53 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>C</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>H3</td>\n",
       "      <td>53</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>HS1376</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>HA_A1011G</td>\n",
       "      <td>0.023458</td>\n",
       "      <td>['S']</td>\n",
       "      <td>[328]</td>\n",
       "      <td>['G']</td>\n",
       "      <td>T316 of HA1</td>\n",
       "      <td>N312 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>312</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>HS1381</td>\n",
       "      <td>2014-12-05</td>\n",
       "      <td>HA_A1486T</td>\n",
       "      <td>0.103901</td>\n",
       "      <td>['Y']</td>\n",
       "      <td>[486]</td>\n",
       "      <td>['F']</td>\n",
       "      <td>Y141 of HA2</td>\n",
       "      <td>Y141 in HA2</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>141</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>HS1425</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>HA_T802C</td>\n",
       "      <td>0.028037</td>\n",
       "      <td>['I']</td>\n",
       "      <td>[258]</td>\n",
       "      <td>['T']</td>\n",
       "      <td>K245 of HA1</td>\n",
       "      <td>V242 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>242</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>HS1294</td>\n",
       "      <td>2014-11-18</td>\n",
       "      <td>HA_G393A</td>\n",
       "      <td>0.066540</td>\n",
       "      <td>['A']</td>\n",
       "      <td>[122]</td>\n",
       "      <td>['T']</td>\n",
       "      <td>E105 of HA1</td>\n",
       "      <td>A106 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>106</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>HS1294</td>\n",
       "      <td>2014-11-18</td>\n",
       "      <td>HA_A801T</td>\n",
       "      <td>0.163603</td>\n",
       "      <td>['I']</td>\n",
       "      <td>[258]</td>\n",
       "      <td>['L']</td>\n",
       "      <td>K245 of HA1</td>\n",
       "      <td>V242 in HA1</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>242</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>HS1284</td>\n",
       "      <td>2014-11-19</td>\n",
       "      <td>HA_G948A</td>\n",
       "      <td>0.048948</td>\n",
       "      <td>['D']</td>\n",
       "      <td>[307]</td>\n",
       "      <td>['N']</td>\n",
       "      <td>S295 of HA1</td>\n",
       "      <td>D291 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>291</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>MH8692</td>\n",
       "      <td>2014-12-27</td>\n",
       "      <td>HA_G891T</td>\n",
       "      <td>0.074946</td>\n",
       "      <td>['A']</td>\n",
       "      <td>[288]</td>\n",
       "      <td>['S']</td>\n",
       "      <td>T276 of HA1</td>\n",
       "      <td>A272 in HA1</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>272</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>MH9045</td>\n",
       "      <td>2015-01-11</td>\n",
       "      <td>HA_G997G</td>\n",
       "      <td>0.160589</td>\n",
       "      <td>['I']</td>\n",
       "      <td>[323]</td>\n",
       "      <td>['R']</td>\n",
       "      <td>K311 of HA1</td>\n",
       "      <td>K307 in HA1</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>307</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>MH5817</td>\n",
       "      <td>2014-01-10</td>\n",
       "      <td>HA_C1372A</td>\n",
       "      <td>0.024559</td>\n",
       "      <td>['T']</td>\n",
       "      <td>[451]</td>\n",
       "      <td>['N']</td>\n",
       "      <td>T107 of HA2</td>\n",
       "      <td>T107 in HA2</td>\n",
       "      <td>cali09</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H1</td>\n",
       "      <td>107</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>MH5546</td>\n",
       "      <td>2013-12-23</td>\n",
       "      <td>HA_G1485A</td>\n",
       "      <td>0.032572</td>\n",
       "      <td>['D']</td>\n",
       "      <td>[489]</td>\n",
       "      <td>['N']</td>\n",
       "      <td>D145 of HA2</td>\n",
       "      <td>D145 in HA2</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H1</td>\n",
       "      <td>145</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>M54062</td>\n",
       "      <td>2011-02-13</td>\n",
       "      <td>HA_C805C</td>\n",
       "      <td>0.024304</td>\n",
       "      <td>['N']</td>\n",
       "      <td>[262]</td>\n",
       "      <td>['T']</td>\n",
       "      <td>T251 of HA1</td>\n",
       "      <td>N248 in HA1</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H1</td>\n",
       "      <td>248</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>MH5405</td>\n",
       "      <td>2013-12-14</td>\n",
       "      <td>HA_G685A</td>\n",
       "      <td>0.020723</td>\n",
       "      <td>['R']</td>\n",
       "      <td>[222]</td>\n",
       "      <td>['K']</td>\n",
       "      <td>R211 of HA1</td>\n",
       "      <td>R208 in HA1</td>\n",
       "      <td>cali09</td>\n",
       "      <td>H1N1</td>\n",
       "      <td>Caton</td>\n",
       "      <td>H1</td>\n",
       "      <td>208</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>MH2936</td>\n",
       "      <td>2013-01-12</td>\n",
       "      <td>HA_C616A</td>\n",
       "      <td>0.024214</td>\n",
       "      <td>['P']</td>\n",
       "      <td>[199]</td>\n",
       "      <td>['Q']</td>\n",
       "      <td>P188 of HA1</td>\n",
       "      <td>P185 in HA1</td>\n",
       "      <td>cali09</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H1</td>\n",
       "      <td>185</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    SPECID       onset   mutation  freq.var Ref_AA AA_pos Var_AA  \\\n",
       "1   M54913  2011-03-30   HA_A258G  0.071215  ['E']   [78]  ['G']   \n",
       "3   MH0778  2012-02-13   HA_T963C  0.071359  ['V']  [313]  ['A']   \n",
       "5   M54724  2011-03-20   HA_C329A  0.087608  ['L']  [102]  ['I']   \n",
       "8   M55001  2011-04-01    HA_T54T  0.032427  ['T']   [10]  ['I']   \n",
       "11  MH2436  2012-12-10   HA_T718C  0.065651  ['I']  [230]  ['T']   \n",
       "13  MH4091  2013-03-05  HA_A1599G  0.079685  ['K']  [524]  ['E']   \n",
       "19  MH8308  2014-12-16  HA_T1526A  0.041457  ['N']  [499]  ['K']   \n",
       "20  MH7851  2014-12-05   HA_T655C  0.174563  ['F']  [209]  ['S']   \n",
       "22  MH7860  2014-12-06   HA_A459G  0.110522  ['T']  [144]  ['A']   \n",
       "28  MH8137  2014-12-15   HA_A855G  0.404932  ['I']  [276]  ['V']   \n",
       "34  MH8052  2014-12-13   HA_G862A  0.030123  ['S']  [278]  ['N']   \n",
       "45  HS1595  2015-01-04   HA_C418T  0.031555  ['S']  [130]  ['L']   \n",
       "50  HS1530  2014-12-22    HA_G75A  0.369782  ['A']   [16]  ['T']   \n",
       "53  HS1530  2014-12-22   HA_A681G  0.360914  ['I']  [218]  ['V']   \n",
       "55  HS1530  2014-12-22   HA_A699A  0.363668  ['G']  [224]  ['R']   \n",
       "57  MH7755  2014-12-03   HA_G564A  0.032187  ['A']  [179]  ['T']   \n",
       "60  MH7796  2014-12-06  HA_G1512A  0.111589  ['G']  [495]  ['R']   \n",
       "61  HS1604  2015-01-11   HA_G997G  0.118966  ['I']  [323]  ['R']   \n",
       "62  MH7611  2014-11-23  HA_G1463A  0.043976  ['M']  [478]  ['I']   \n",
       "64  HS1455  2014-12-13  HA_A1381A  0.350988  ['R']  [451]  ['H']   \n",
       "65  HS1450  2014-12-13   HA_G644T  0.037827  ['K']  [205]  ['N']   \n",
       "68  HS1473  2014-12-15  HA_T1071C  0.177116  ['F']  [348]  ['L']   \n",
       "69  HS1401  2014-12-06   HA_A879C  0.034116  ['M']  [284]  ['L']   \n",
       "70  HS1376  2014-12-03   HA_C236A  0.025191  ['D']   [69]  ['E']   \n",
       "71  HS1376  2014-12-03  HA_A1011G  0.023458  ['S']  [328]  ['G']   \n",
       "72  HS1381  2014-12-05  HA_A1486T  0.103901  ['Y']  [486]  ['F']   \n",
       "73  HS1425  2014-12-06   HA_T802C  0.028037  ['I']  [258]  ['T']   \n",
       "74  HS1294  2014-11-18   HA_G393A  0.066540  ['A']  [122]  ['T']   \n",
       "75  HS1294  2014-11-18   HA_A801T  0.163603  ['I']  [258]  ['L']   \n",
       "78  HS1284  2014-11-19   HA_G948A  0.048948  ['D']  [307]  ['N']   \n",
       "79  MH8692  2014-12-27   HA_G891T  0.074946  ['A']  [288]  ['S']   \n",
       "80  MH9045  2015-01-11   HA_G997G  0.160589  ['I']  [323]  ['R']   \n",
       "85  MH5817  2014-01-10  HA_C1372A  0.024559  ['T']  [451]  ['N']   \n",
       "88  MH5546  2013-12-23  HA_G1485A  0.032572  ['D']  [489]  ['N']   \n",
       "93  M54062  2011-02-13   HA_C805C  0.024304  ['N']  [262]  ['T']   \n",
       "95  MH5405  2013-12-14   HA_G685A  0.020723  ['R']  [222]  ['K']   \n",
       "99  MH2936  2013-01-12   HA_C616A  0.024214  ['P']  [199]  ['Q']   \n",
       "\n",
       "             PDB_4JTV           PDB_4HMG       run Antigenic Source subtype  \\\n",
       "1         G59 of HA1         I62 in HA1    perth_2         E  Smith      H3   \n",
       "3        I301 of HA1        V297 in HA1    perth_2         C    Lee      H3   \n",
       "5         Y84 of HA1         L86 in HA1      perth         E    Lee      H3   \n",
       "8   an alignment gap   an alignment gap      perth        NA     NA      H3   \n",
       "11       K217 of HA1        I214 in HA1        vic         D    Lee      H3   \n",
       "13  an alignment gap   an alignment gap        vic        NA     NA      H3   \n",
       "19       N154 of HA2        N154 in HA2       HK_2        NA     NA      H3   \n",
       "20       S196 of HA1        S193 in HA1       HK_2         B  Wiley      H3   \n",
       "22       P130 of HA1        T128 in HA1       HK_8         B    Lee      H3   \n",
       "28       M263 of HA1        M260 in HA1       HK_2         E  Smith      H3   \n",
       "34       N266 of HA1        T262 in HA1       HK_8         E  Smith      H3   \n",
       "45       S116 of HA1        S114 in HA1       HK_7        NA     NA      H3   \n",
       "50  an alignment gap   an alignment gap       HK_1        NA     NA      H3   \n",
       "53       V205 of HA1        V202 in HA1       HK_1        NA     NA      H3   \n",
       "55       R211 of HA1        R208 in HA1       HK_1         D    Lee      H3   \n",
       "57       K166 of HA1        V163 in HA1       HK_7         B    Lee      H3   \n",
       "60       E150 of HA2        E150 in HA2       HK_7        NA     NA      H3   \n",
       "61       K311 of HA1        K307 in HA1       HK_7         C    Lee      H3   \n",
       "62       I133 of HA2        M133 in HA2       HK_7        NA     NA      H3   \n",
       "64       R106 of HA2        H106 in HA2       HK_1        NA     NA      H3   \n",
       "65       A192 of HA1        Q189 in HA1       HK_6         B  Wiley      H3   \n",
       "68         F3 of HA2          F3 in HA2       HK_6        NA     NA      H3   \n",
       "69       I272 of HA1        M268 in HA1       HK_6        NA     NA      H3   \n",
       "70        K49 of HA1         N53 in HA1       HK_6         C  Wiley      H3   \n",
       "71       T316 of HA1        N312 in HA1       HK_6         C    Lee      H3   \n",
       "72       Y141 of HA2        Y141 in HA2       HK_6        NA     NA      H3   \n",
       "73       K245 of HA1        V242 in HA1       HK_6         D    Lee      H3   \n",
       "74       E105 of HA1        A106 in HA1       HK_6        NA     NA      H3   \n",
       "75       K245 of HA1        V242 in HA1       HK_1         D    Lee      H3   \n",
       "78       S295 of HA1        D291 in HA1       HK_6        NA     NA      H3   \n",
       "79       T276 of HA1        A272 in HA1       HK_1        NA     NA      H3   \n",
       "80       K311 of HA1        K307 in HA1       HK_1         C    Lee      H3   \n",
       "85       T107 of HA2        T107 in HA2     cali09        NA     NA      H1   \n",
       "88       D145 of HA2        D145 in HA2   cali09_2        NA     NA      H1   \n",
       "93       T251 of HA1        N248 in HA1   cali09_2        NA     NA      H1   \n",
       "95       R211 of HA1        R208 in HA1     cali09      H1N1  Caton      H1   \n",
       "99       P188 of HA1        P185 in HA1     cali09        NA     NA      H1   \n",
       "\n",
       "   H3_pos H3_seg  \n",
       "1      62    HA1  \n",
       "3     297    HA1  \n",
       "5      86    HA1  \n",
       "8       -      -  \n",
       "11    214    HA1  \n",
       "13      -      -  \n",
       "19    154    HA2  \n",
       "20    193    HA1  \n",
       "22    128    HA1  \n",
       "28    260    HA1  \n",
       "34    262    HA1  \n",
       "45    114    HA1  \n",
       "50      -      -  \n",
       "53    202    HA1  \n",
       "55    208    HA1  \n",
       "57    163    HA1  \n",
       "60    150    HA2  \n",
       "61    307    HA1  \n",
       "62    133    HA2  \n",
       "64    106    HA2  \n",
       "65    189    HA1  \n",
       "68      3    HA2  \n",
       "69    268    HA1  \n",
       "70     53    HA1  \n",
       "71    312    HA1  \n",
       "72    141    HA2  \n",
       "73    242    HA1  \n",
       "74    106    HA1  \n",
       "75    242    HA1  \n",
       "78    291    HA1  \n",
       "79    272    HA1  \n",
       "80    307    HA1  \n",
       "85    107    HA2  \n",
       "88    145    HA2  \n",
       "93    248    HA1  \n",
       "95    208    HA1  \n",
       "99    185    HA1  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HA_nonsyn[[\"SPECID\",\"onset\",\"mutation\",\"freq.var\",\"Ref_AA\",\"AA_pos\",\"Var_AA\",\"PDB_4JTV\",\"PDB_4HMG\",\"run\",\"Antigenic\",\"Source\",\"subtype\",\"H3_pos\",\"H3_seg\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Antigenic sites\n",
    "\n",
    "Here are the antigenic sites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "AA = HA_nonsyn.loc[HA_nonsyn[\"Antigenic\"]!=\"NA\"]\n",
    "len(AA)\n",
    "HA_nonsyn.to_csv(\"./HA_nonsynom.csv\")\n",
    "AA[[\"HOUSE_ID\",\"ENROLLID\",\"SPECID\",\"onset\",\"mutation\",\"freq.var\",\"Ref_AA\",\"H3_pos\",\"Var_AA\",\"PDB_4HMG\",\"PDB_4JTV\",\"pcr_result\",\"Antigenic\",\"Source\",\"vaccination_status\",\"DPI\"]].to_csv(\"./antigenic_isnv.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HOUSE_ID</th>\n",
       "      <th>ENROLLID</th>\n",
       "      <th>SPECID</th>\n",
       "      <th>onset</th>\n",
       "      <th>mutation</th>\n",
       "      <th>freq.var</th>\n",
       "      <th>Ref_AA</th>\n",
       "      <th>H3_pos</th>\n",
       "      <th>Var_AA</th>\n",
       "      <th>PDB_4HMG</th>\n",
       "      <th>PDB_4JTV</th>\n",
       "      <th>pcr_result</th>\n",
       "      <th>Antigenic</th>\n",
       "      <th>Source</th>\n",
       "      <th>vaccination_status</th>\n",
       "      <th>DPI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1111</td>\n",
       "      <td>300481</td>\n",
       "      <td>M54913</td>\n",
       "      <td>2011-03-30</td>\n",
       "      <td>HA_A258G</td>\n",
       "      <td>0.071215</td>\n",
       "      <td>['E']</td>\n",
       "      <td>62</td>\n",
       "      <td>['G']</td>\n",
       "      <td>I62 in HA1</td>\n",
       "      <td>G59 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>E</td>\n",
       "      <td>Smith</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2166</td>\n",
       "      <td>320661</td>\n",
       "      <td>MH0778</td>\n",
       "      <td>2012-02-13</td>\n",
       "      <td>HA_T963C</td>\n",
       "      <td>0.071359</td>\n",
       "      <td>['V']</td>\n",
       "      <td>297</td>\n",
       "      <td>['A']</td>\n",
       "      <td>V297 in HA1</td>\n",
       "      <td>I301 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1302</td>\n",
       "      <td>301355</td>\n",
       "      <td>M54724</td>\n",
       "      <td>2011-03-20</td>\n",
       "      <td>HA_C329A</td>\n",
       "      <td>0.087608</td>\n",
       "      <td>['L']</td>\n",
       "      <td>86</td>\n",
       "      <td>['I']</td>\n",
       "      <td>L86 in HA1</td>\n",
       "      <td>Y84 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>E</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3075</td>\n",
       "      <td>331045</td>\n",
       "      <td>MH2436</td>\n",
       "      <td>2012-12-10</td>\n",
       "      <td>HA_T718C</td>\n",
       "      <td>0.065651</td>\n",
       "      <td>['I']</td>\n",
       "      <td>214</td>\n",
       "      <td>['T']</td>\n",
       "      <td>I214 in HA1</td>\n",
       "      <td>K217 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>5219</td>\n",
       "      <td>50935</td>\n",
       "      <td>MH7851</td>\n",
       "      <td>2014-12-05</td>\n",
       "      <td>HA_T655C</td>\n",
       "      <td>0.174563</td>\n",
       "      <td>['F']</td>\n",
       "      <td>193</td>\n",
       "      <td>['S']</td>\n",
       "      <td>S193 in HA1</td>\n",
       "      <td>S196 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>B</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>5263</td>\n",
       "      <td>51106</td>\n",
       "      <td>MH7860</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>HA_A459G</td>\n",
       "      <td>0.110522</td>\n",
       "      <td>['T']</td>\n",
       "      <td>128</td>\n",
       "      <td>['A']</td>\n",
       "      <td>T128 in HA1</td>\n",
       "      <td>P130 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>B</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>5290</td>\n",
       "      <td>51225</td>\n",
       "      <td>MH8137</td>\n",
       "      <td>2014-12-15</td>\n",
       "      <td>HA_A855G</td>\n",
       "      <td>0.404932</td>\n",
       "      <td>['I']</td>\n",
       "      <td>260</td>\n",
       "      <td>['V']</td>\n",
       "      <td>M260 in HA1</td>\n",
       "      <td>M263 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>E</td>\n",
       "      <td>Smith</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>5302</td>\n",
       "      <td>51273</td>\n",
       "      <td>MH8052</td>\n",
       "      <td>2014-12-13</td>\n",
       "      <td>HA_G862A</td>\n",
       "      <td>0.030123</td>\n",
       "      <td>['S']</td>\n",
       "      <td>262</td>\n",
       "      <td>['N']</td>\n",
       "      <td>T262 in HA1</td>\n",
       "      <td>N266 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>E</td>\n",
       "      <td>Smith</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>5098</td>\n",
       "      <td>50419</td>\n",
       "      <td>HS1530</td>\n",
       "      <td>2014-12-22</td>\n",
       "      <td>HA_A699A</td>\n",
       "      <td>0.363668</td>\n",
       "      <td>['G']</td>\n",
       "      <td>208</td>\n",
       "      <td>['R']</td>\n",
       "      <td>R208 in HA1</td>\n",
       "      <td>R211 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>5033</td>\n",
       "      <td>50141</td>\n",
       "      <td>MH7755</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>HA_G564A</td>\n",
       "      <td>0.032187</td>\n",
       "      <td>['A']</td>\n",
       "      <td>163</td>\n",
       "      <td>['T']</td>\n",
       "      <td>V163 in HA1</td>\n",
       "      <td>K166 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>B</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>5034</td>\n",
       "      <td>50143</td>\n",
       "      <td>HS1604</td>\n",
       "      <td>2015-01-11</td>\n",
       "      <td>HA_G997G</td>\n",
       "      <td>0.118966</td>\n",
       "      <td>['I']</td>\n",
       "      <td>307</td>\n",
       "      <td>['R']</td>\n",
       "      <td>K307 in HA1</td>\n",
       "      <td>K311 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>5289</td>\n",
       "      <td>51220</td>\n",
       "      <td>HS1450</td>\n",
       "      <td>2014-12-13</td>\n",
       "      <td>HA_G644T</td>\n",
       "      <td>0.037827</td>\n",
       "      <td>['K']</td>\n",
       "      <td>189</td>\n",
       "      <td>['N']</td>\n",
       "      <td>Q189 in HA1</td>\n",
       "      <td>A192 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>B</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>5033</td>\n",
       "      <td>50141</td>\n",
       "      <td>HS1376</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>HA_C236A</td>\n",
       "      <td>0.025191</td>\n",
       "      <td>['D']</td>\n",
       "      <td>53</td>\n",
       "      <td>['E']</td>\n",
       "      <td>N53 in HA1</td>\n",
       "      <td>K49 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>C</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>5033</td>\n",
       "      <td>50141</td>\n",
       "      <td>HS1376</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>HA_A1011G</td>\n",
       "      <td>0.023458</td>\n",
       "      <td>['S']</td>\n",
       "      <td>312</td>\n",
       "      <td>['G']</td>\n",
       "      <td>N312 in HA1</td>\n",
       "      <td>T316 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>5269</td>\n",
       "      <td>51132</td>\n",
       "      <td>HS1425</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>HA_T802C</td>\n",
       "      <td>0.028037</td>\n",
       "      <td>['I']</td>\n",
       "      <td>242</td>\n",
       "      <td>['T']</td>\n",
       "      <td>V242 in HA1</td>\n",
       "      <td>K245 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>5147</td>\n",
       "      <td>50630</td>\n",
       "      <td>HS1294</td>\n",
       "      <td>2014-11-18</td>\n",
       "      <td>HA_A801T</td>\n",
       "      <td>0.163603</td>\n",
       "      <td>['I']</td>\n",
       "      <td>242</td>\n",
       "      <td>['L']</td>\n",
       "      <td>V242 in HA1</td>\n",
       "      <td>K245 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>5034</td>\n",
       "      <td>50143</td>\n",
       "      <td>MH9045</td>\n",
       "      <td>2015-01-11</td>\n",
       "      <td>HA_G997G</td>\n",
       "      <td>0.160589</td>\n",
       "      <td>['I']</td>\n",
       "      <td>307</td>\n",
       "      <td>['R']</td>\n",
       "      <td>K307 in HA1</td>\n",
       "      <td>K311 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>4185</td>\n",
       "      <td>UM40738</td>\n",
       "      <td>MH5405</td>\n",
       "      <td>2013-12-14</td>\n",
       "      <td>HA_G685A</td>\n",
       "      <td>0.020723</td>\n",
       "      <td>['R']</td>\n",
       "      <td>208</td>\n",
       "      <td>['K']</td>\n",
       "      <td>R208 in HA1</td>\n",
       "      <td>R211 of HA1</td>\n",
       "      <td>A/H1N1</td>\n",
       "      <td>H1N1</td>\n",
       "      <td>Caton</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    HOUSE_ID ENROLLID  SPECID       onset   mutation  freq.var Ref_AA H3_pos  \\\n",
       "1       1111   300481  M54913  2011-03-30   HA_A258G  0.071215  ['E']     62   \n",
       "3       2166   320661  MH0778  2012-02-13   HA_T963C  0.071359  ['V']    297   \n",
       "5       1302   301355  M54724  2011-03-20   HA_C329A  0.087608  ['L']     86   \n",
       "11      3075   331045  MH2436  2012-12-10   HA_T718C  0.065651  ['I']    214   \n",
       "20      5219    50935  MH7851  2014-12-05   HA_T655C  0.174563  ['F']    193   \n",
       "22      5263    51106  MH7860  2014-12-06   HA_A459G  0.110522  ['T']    128   \n",
       "28      5290    51225  MH8137  2014-12-15   HA_A855G  0.404932  ['I']    260   \n",
       "34      5302    51273  MH8052  2014-12-13   HA_G862A  0.030123  ['S']    262   \n",
       "55      5098    50419  HS1530  2014-12-22   HA_A699A  0.363668  ['G']    208   \n",
       "57      5033    50141  MH7755  2014-12-03   HA_G564A  0.032187  ['A']    163   \n",
       "61      5034    50143  HS1604  2015-01-11   HA_G997G  0.118966  ['I']    307   \n",
       "65      5289    51220  HS1450  2014-12-13   HA_G644T  0.037827  ['K']    189   \n",
       "70      5033    50141  HS1376  2014-12-03   HA_C236A  0.025191  ['D']     53   \n",
       "71      5033    50141  HS1376  2014-12-03  HA_A1011G  0.023458  ['S']    312   \n",
       "73      5269    51132  HS1425  2014-12-06   HA_T802C  0.028037  ['I']    242   \n",
       "75      5147    50630  HS1294  2014-11-18   HA_A801T  0.163603  ['I']    242   \n",
       "80      5034    50143  MH9045  2015-01-11   HA_G997G  0.160589  ['I']    307   \n",
       "95      4185  UM40738  MH5405  2013-12-14   HA_G685A  0.020723  ['R']    208   \n",
       "\n",
       "   Var_AA      PDB_4HMG      PDB_4JTV pcr_result Antigenic Source  \\\n",
       "1   ['G']   I62 in HA1    G59 of HA1      A/H3N2         E  Smith   \n",
       "3   ['A']  V297 in HA1   I301 of HA1      A/H3N2         C    Lee   \n",
       "5   ['I']   L86 in HA1    Y84 of HA1      A/H3N2         E    Lee   \n",
       "11  ['T']  I214 in HA1   K217 of HA1      A/H3N2         D    Lee   \n",
       "20  ['S']  S193 in HA1   S196 of HA1      A/H3N2         B  Wiley   \n",
       "22  ['A']  T128 in HA1   P130 of HA1      A/H3N2         B    Lee   \n",
       "28  ['V']  M260 in HA1   M263 of HA1      A/H3N2         E  Smith   \n",
       "34  ['N']  T262 in HA1   N266 of HA1      A/H3N2         E  Smith   \n",
       "55  ['R']  R208 in HA1   R211 of HA1      A/H3N2         D    Lee   \n",
       "57  ['T']  V163 in HA1   K166 of HA1      A/H3N2         B    Lee   \n",
       "61  ['R']  K307 in HA1   K311 of HA1      A/H3N2         C    Lee   \n",
       "65  ['N']  Q189 in HA1   A192 of HA1      A/H3N2         B  Wiley   \n",
       "70  ['E']   N53 in HA1    K49 of HA1      A/H3N2         C  Wiley   \n",
       "71  ['G']  N312 in HA1   T316 of HA1      A/H3N2         C    Lee   \n",
       "73  ['T']  V242 in HA1   K245 of HA1      A/H3N2         D    Lee   \n",
       "75  ['L']  V242 in HA1   K245 of HA1      A/H3N2         D    Lee   \n",
       "80  ['R']  K307 in HA1   K311 of HA1      A/H3N2         C    Lee   \n",
       "95  ['K']  R208 in HA1   R211 of HA1      A/H1N1      H1N1  Caton   \n",
       "\n",
       "    vaccination_status  DPI  \n",
       "1                    0    0  \n",
       "3                    1    1  \n",
       "5                    1    1  \n",
       "11                   1    1  \n",
       "20                   0    3  \n",
       "22                   1    3  \n",
       "28                   1    1  \n",
       "34                   1    0  \n",
       "55                   1    4  \n",
       "57                   1    2  \n",
       "61                   1    1  \n",
       "65                   1   -1  \n",
       "70                   1    1  \n",
       "71                   1    1  \n",
       "73                   1    2  \n",
       "75                   1    1  \n",
       "80                   1    2  \n",
       "95                   0    2  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AA[[\"HOUSE_ID\",\"ENROLLID\",\"SPECID\",\"onset\",\"mutation\",\"freq.var\",\"Ref_AA\",\"H3_pos\",\"Var_AA\",\"PDB_4HMG\",\"PDB_4JTV\",\"pcr_result\",\"Antigenic\",\"Source\",\"vaccination_status\",\"DPI\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "Here are the frequencies of the antigenic mutations against day post symptom onset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAuEAAAIACAYAAAAsWLK1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xl4VPWh//HPbNmYmSSEYQsSBEMAowjEogIioO0VoaBi\nRWSplBYtbbWLC6WtVvRab937tPcWaxXQul4tqIi2QGMRq0K1QikQFQIE1GBIJvvCnN8fXObXEZFB\ncr4nc3i/nsfnYSYnk898XPh4ODnxWJZlCQAAAIAxXqcDAAAAACcaRjgAAABgGCMcAAAAMIwRDgAA\nABjGCAcAAAAMY4QDAAAAhvmdDmCHsrIyrVy5UpZlaejQoRo5cmTCx3fs2KHHH39cubm5kqSBAwdq\n9OjRTkQFAADACch1IzwWi2nFihWaNWuWQqGQFi1apKKiIkUikYTjCgoKNG3aNIdSAgAA4ETmustR\nKioqlJeXp5ycHPl8PhUXF2vr1q1OxwIAAADiXDfCa2trFQ6H44/D4bCi0ehhx+3atUv//d//rcce\ne0wff/yxyYgAAAA4wbnucpRk9OjRQ9///veVlpamsrIyPfHEE/re974nSYpGo6qrq0s4PhgMJgx7\nAAAA4Hi4boSHQiHV1NTEH0ej0cMGdHp6evzXhYWFevHFF9XQ0KCsrCxt2LBBpaWlCcePHj1aY8aM\nsTc4AAAAThiuG+H5+fmqqqpSdXW1gsGgNm3apClTpiQcU1dXp2AwKEnavXu3LMtSVlaWJGnYsGEq\nKipKOD4YDGr//v1qa2sz8yaOU3p6upqbm52OkRS/36/c3Fz6tQHd2ot+7UW/9kq1funWXqnYrxu4\nboR7vV6NHz9eS5culWVZGjJkiCKRiNavXy9JKikp0ebNm/XWW2/J5/PJ7/frsssui39+OBz+zEtP\nKisr1draaux9HA+/358yWQ9pa2tLmcyp1i/d2ot+7UW/9kqVfunWXqnYrxu4boRLBy8xKSwsTHiu\npKQk/usvfelL+tKXvmQ6FgAAACDJhXdHAQAAADo6RjgAAABgGCMcAAAAMIwRDgAAABjGCAcAAAAM\nY4QDAAAAhjHCAQAAAMMY4QAAAIBhjHAAAADAMEY4AAAAYBgjHAAAADCMEQ4AAAAYxggHAAAADGOE\nAwAAAIYxwgEAAADDGOEAAACAYYxwAAAAwDBGOAAAAGAYIxwAAAAwjBEOAAAAGMYIBwAAAAxjhAMA\nAACGMcIBAAAAwxjhAAAAgGGMcAAAAMAwRjgAAABgGCMcAAAAMIwRDgAAABjGCAcAAAAMY4QDAAAA\nhjHCAQAAAMMY4QAAAIBhjHAAAADAMI9lWZbTITq6pqYmNTU1KVWq8nq9isViTsdIisfjUVpamlpa\nWui3ndGtvejXXvRrr1Trl27tlWr95uTkOB2jXfidDpAKMjIyVFtbq9bWVqejJCUzM1ONjY1Ox0hK\nIBBQTk6O6uvr6bed0a296Nde9GuvVOuXbu2Vav26BZejAAAAAIYxwgEAAADDGOEAAACAYYxwAAAA\nwDBGOAAAAGAYIxwAAAAwjBEOAAAAGMYIBwAAAAxjhAMAAACGMcIBAAAAwxjhAAAAgGGMcAAAAMAw\nRjgAAABgGCMcAAAAMIwRDgAAABjGCAcAAAAMY4QDAAAAhjHCAQAAAMMY4QAAAIBhjHAAAADAMEY4\nAAAAYBgjHAAAADCMEQ4AAAAYxggHAAAADGOEAwAAAIYxwgEAAADDGOEAAACAYYxwAAAAwDBGOAAA\nAGCY3+kAAJCKNm7cKI/Ho8LCQnm9nM8AABwbfucAgGP0y1/+UuPGjdPYsWM1efJkNTY2Oh0JAJBi\nXDnCy8rK9Ktf/UoPPPCA1q5de8TjKioqdOutt2rz5s0G0wFIZc3NzXrggQfijzds2KA///nPDiYC\nAKQi143wWCymFStWaMaMGZo3b542btyoysrKzzzuz3/+s/r16+dASgCpyufzye9PvJIvPT3doTQA\ngFTluhFeUVGhvLw85eTkyOfzqbi4WFu3bj3suDfffFODBg1Sp06dHEgJIFX5/X7dcccd8SE+ceJE\nnX/++Q6nAgCkGteN8NraWoXD4fjjcDisaDSacEw0GtWWLVt05plnmo4HwAWmTp2qzZs3a+fOnXro\noYf4xkwAwDE7Ie+OsnLlyiOeuYpGo6qrq0t4LhgMHvbHzx2Zz+dTIBBwOkZSDvVKv+2Pbu3VpUsX\n5ebmav/+/U5HSVoq9cs/v/ZKtX7p1l6p2K8buOed/J9QKKSampr442g0mnBmXJL27NmjZ555RpLU\n0NCgsrIyeb1eDRgwQBs2bFBpaWnC8TNnzlT37t1T6rrPVPqHtKGhQenp6fRrA7q1F/3ai37tlWr9\n0q29Uqlft3Bd4/n5+aqqqlJ1dbWCwaA2bdqkKVOmJBxz3XXXxX/9xz/+Uf3799eAAQMkScOGDVNR\nUVHC8cFgUM3NzWpra7P/DbSD9PR0NTc3Ox0jKX6/P342kX7bF93ai37tRb/2SrV+6dZeqdZvVlaW\n0zHahetGuNfr1fjx47V06VJZlqUhQ4YoEolo/fr1kqSSkpLP/fxwOHzYmXNJqqysVGtrqy2Z25vf\n70+ZrIe0tbWlTOZU65du7UW/9qJfe6VKv3Rrr1Ts1w1cN8IlqbCwUIWFhQnPHWl8T5482UQkAAAA\nII5v6QcAAAAMY4QDAAAAhjHCAQAAAMMY4QAAAIBhjHAAAADAMEY4AAAAYBgjHAAAADCMEQ4AAAAY\nxggHAAAADGOEAwAAAIYxwgEAAADDGOEAAACAYYxwAAAAwDBGOAAAAGAYIxwAAAAwjBEOAAAAGMYI\nBwAAAAxjhAMAAACGMcIBAAAAwxjhAAAAgGF+pwMAAAC4QWNjo37zm9+orq5OEyZM0KBBg5yOhA6M\nEQ4AANAO5s2bp5dfflmS9OCDD+qll17SKaec4nAqdFRcjgIAAHCcLMvSqlWr4o8bGhr02muvOZgI\nHR0jHAAA4Dh5PB7169cv4TnOguPzMMIBAADawe9+9zuNGjVKp512mu644w6NGDHC6UjowLgmHAAA\noB307dtX//u//6tIJKLKykq1trY6HQkdGGfCAQAAAMMY4QAAAIBhjHAAAADAMEY4AAAAYBgjHAAA\nADCMEQ4AAAAYxggHAAAADPNYlmU5HaKja2pqUlNTk1KlKq/Xq1gs5nSMpHg8HqWlpamlpYV+2xnd\n2ot+7UW/9kq1funWXqnWb05OjtMx2gU/rCcJGRkZqq2tTZmb7mdmZqqxsdHpGEkJBALKyclRfX09\n/bYzurUX/dqLfu2Vav3Srb1SrV+34HIUAAAAwDBGOAAAAGAYIxwAAAAwjBEOAAAAGMYIBwAAAAxj\nhAMAAACGMcIBAAAAwxjhAAAAgGGMcAAAAMAwRjgAAABgGCMcAAAAMIwRDgAAABjGCAcAAAAMY4QD\nAAAAhjHCAQAAAMMY4QAAAIBhjHAAAADAMEY4AAAAYBgjHAAAADCMEQ4AAAAYxggHAAAADGOEAwAA\nAIYxwgEAAADDGOEAAACAYYxwAAAAwDBGOAAAAGAYIxwAAAAwjBEOAAAAGMYIBwAAAAxjhAMAAACG\n+Z0OYIeysjKtXLlSlmVp6NChGjlyZMLHt2zZojVr1sjj8cjj8eiCCy5Q3759HUoLAACAE43rRngs\nFtOKFSs0a9YshUIhLVq0SEVFRYpEIvFj+vbtqwEDBkiSPvroIz3xxBO69tprnYoMAACAE4zrLkep\nqKhQXl6ecnJy5PP5VFxcrK1btyYck5aWFv91S0uLsrKyTMcEAADACcx1Z8Jra2sVDofjj8PhsCoq\nKg477l//+pdWrVqluro6TZ8+3WREAAAAnOBcN8KTNXDgQA0cOFDl5eV67rnn9N3vfleSFI1GVVdX\nl3BsMBiU3586Vfl8PgUCAadjJOVQr/Tb/ujWXvRrL/q1V6r1S7f2SsV+3cA97+T/hEIh1dTUxB9H\no9GEM+OfVlBQoFgspoaGBmVlZWnDhg0qLS1NOGb06NEaM2aMbZkh5ebmOh3BtejWXvRrL/q1F/3a\nh25xNK4b4fn5+aqqqlJ1dbWCwaA2bdqkKVOmJBxTVVWlzp07S5L27NkjSfHrwocNG6aioqKE44PB\noPbv36+2tjYD7+D4paenq7m52ekYSfH7/crNzaVfG9CtvejXXvRrr1Trl27tlYr9uoHrRrjX69X4\n8eO1dOlSWZalIUOGKBKJaP369ZKkkpISbd68Wf/4xz/k8/mUlpaWMNLD4fBnnjmvrKxUa2ursfdx\nPPx+f8pkPaStrS1lMqdav3RrL/q1F/3aK1X6pVt7pWK/buC6ES5JhYWFKiwsTHiupKQk/uuRI0ce\ndu9wAAAAwBTX3aIQAAAA6OgY4QAAAIBhjHAAAADAMEY4AAAAYBgjHAAAADCMEQ4AAAAYxggHAAAA\nDGOEAwAAAIYxwgEAAADDGOEAAACAYYxwAAAAwDBGOAAAAGAYIxwAAAAwjBEOAAAAGMYIBwAAAAxj\nhAMAAACGMcIBAAAAwxjhAAAAgGGMcAAAAMAwRjgAAABgGCMcAAAAMIwRDgAAABjGCAcAAAAMY4QD\nAAAAhjHCAQAAAMMY4QAAAIBhjHAAAADAMEY4AAAAYBgjHAAAADCMEQ4AAAAYxggHAAAADGOEAwAA\nAIYxwgEAAADDPJZlWU6H6OiamprU1NSkVKnK6/UqFos5HSMpHo9HaWlpamlpod92Rrf2ol970a+9\nUq1furVXqvWbk5PjdIx24Xc6QCrIyMhQbW2tWltbnY6SlMzMTDU2NjodIymBQEA5OTmqr6+n33ZG\nt/aiX3vRr71SrV+6tVeq9esWXI4CAAAAGMYIBwAAAAxjhAMAAACGMcIBAAAAwxjhAAAAgGGMcAAA\nAMAwRjgAAABgGCMcAAAAMIwRDgAAABjGCAcAAAAMY4QDAAAAhjHCAQAAAMMY4QAAAIBhjHAAAADA\nMEY4AAAAYBgjHAAAADCMEQ4AAAAYxggHAAAADGOEAwAAAIYxwgEAAADDGOEAAACAYYxwAAAAwDBG\nOAAAAGAYIxwAAAAwjBEOAAAAGMYIBwAAAAxjhAMAAACGMcIBAAAAwxjhAAAAgGGMcAAAAMAwv9MB\n7FBWVqaVK1fKsiwNHTpUI0eOTPj4u+++q9dee02SlJaWpgkTJqhbt25ORAUAAMAJyHUjPBaLacWK\nFZo1a5ZCoZAWLVqkoqIiRSKR+DG5ubm66qqrlJGRobKyMi1fvlzf/OY3HUwNAACAE4nrLkepqKhQ\nXl6ecnJy5PP5VFxcrK1btyYcc9JJJykjI0OS1KtXL9XW1joRFQAAACco143w2tpahcPh+ONwOKxo\nNHrE4//+97/rlFNOMRENAAAAkOTCy1GOxfbt2/XOO+9o9uzZ8eei0ajq6uoSjgsGg/L7U6cqn8+n\nQCDgdIykHOqVftsf3dqLfu1Fv/ZKtX7p1l6p2K8buOed/J9QKKSampr442g0mnBm/JAPP/xQzz//\nvKZPn67MzMz48xs2bFBpaWnCsTNnzlT37t2Vnp5uX/B2lkr/kDY0NCg9PZ1+bUC39qJfe9GvvVKt\nX7q1Vyr16xZJN37//ffryiuvVJcuXezMc9zy8/NVVVWl6upqBYNBbdq0SVOmTEk4prq6Wk899ZQu\nvvhide7cOeFjw4YNU1FRUcJzwWBQzc3Namtrsz1/e0hPT1dzc7PTMZLi9/uVm5ur/fv30287o1t7\n0a+96NdeqdYv3dor1frNyspyOka7SHqEr169WgsWLNB5552nGTNmaPLkyR3y//C8Xq/Gjx+vpUuX\nyrIsDRkyRJFIROvXr5cklZSU6NVXX1VjY6NefPHF+Od861vfknTwGvLPOnNeWVmp1tZWc2/kOPj9\n/pTJekhbW1vKZE61funWXvRrL/q1V6r0S7f2SsV+3SDpEb5s2TJ98skneuKJJ3Tffffp6quv1qWX\nXqqZM2fq3HPPtTPjMSssLFRhYWHCcyUlJfFff/WrX9VXv/pV07EAAAAAScd4d5S8vDzNmzdPr7/+\nukpLS/XWW29pzJgx6tOnj26//fbDvqERAAAAwOGO+RaFq1at0lVXXaXzzjtPXbt21eLFi/Xoo4/q\n7bff1oUXXmhHRgAAAMBVkr4c5Uc/+pGeeOIJZWdna8aMGdq4caPy8/PjHx8+fLhyc3NtCQkAAAC4\nSdIjvKmpSc8995zOPPPMz/x4IBCIf/MjAAAAgCNLeoTPnz//sFvC7N+/X42NjerZs6ckacCAAe2b\nDgAAAHChpK8Jnzx5snbv3p3w3O7du3XxxRe3eygAAADAzZIe4Vu3btVpp52W8Nxpp52mLVu2tHso\nAAAAwM2SHuFdu3bVe++9l/Dce++9p7y8vHYPBQAAALhZ0iN89uzZuvTSS/XCCy9o8+bNev755zVl\nyhTNmTPHznwAAACA6yT9jZk33XSTAoGAfvSjH2nXrl066aSTNGfOHP3gBz+wMx8AAADgOkmPcK/X\nq+uvv17XX3+9nXkAAAAA10t6hEsHvznzH//4x2E/nn727NntGgoAAABws6RH+H/+53/q1ltv1eDB\ngxPuF+7xeBjhAAAAwDFIeoTfd999evPNN3X66afbmQcAANiksbFRN9xwg9566y0NHz5cd955pzIy\nMpyOBZyQkh7hmZmZ/ERMAABS2L333qtnn31WkrRr1y4Fg0HdfvvtDqcCTkxJ36Jw4cKF+u53v6u9\ne/cqFosl/AUAADq+HTt2fO5jAOYkPcK//vWv68EHH1SvXr0UCAQUCATk9/sVCATszAcAANrJ+PHj\nEx5fdNFFDiUBkPTlKNu3b7czBwAAsNnkyZMVCoW0YcMGjR49WiNHjlRra6vTsYATUtIjvKCgQJIU\ni8X00UcfqUePHraFAgAA9hg3bpz+4z/+Q5FIRJWVlU7HAU5YSV+OUl1drWnTpikjI0OnnHKKJGn5\n8uX6yU9+Yls4AAAAwI2SHuFXX321srOzVV5errS0NEnS2WefrSeffNK2cAAAAIAbJX05yqpVq7Rn\nzx4FAgF5PB5JUiQS0ccff2xbOADoiCoqKvSLX/xCTU1NmjFjhs4991ynIwEAUkzSIzw7O1v79u1L\nuBZ8586dXBsO4IQzffp0bdu2TZL0pz/9SS+//LKKioocTgUASCVJX44yZ84cXXrppVqzZo1isZhe\nf/11zZo1S1dffbWd+QCgQ2loaIgPcElqbW3Vpk2bHEzkPhUVFXrggQe0aNEi7twBwLWSPhN+4403\nKjMzU/PmzVNra6tmz56tuXPn6tprr7UzHwB0KFlZWRo0aJA2b94sSUpPT9fgwYMdTuUeH3/8sSZM\nmBC/1PHCCy/U7373O4dTAUD7S3qEezweXXvttYxuACe8xx57TPfcc4+ampp0xRVXxO8YheO3du3a\nhO81eumll9TY2KjMzEwHUwFA+0t6hK9evfqIHxs7dmy7hAGAVNC1a1fdfffd8fssc8lE+8nPz094\n3KVLF2VkZDiUBgDsk/QI/8Y3vpHwuLKyUi0tLerVq5c++OCDdg8GADjxDB8+XD/+8Y/14IMPKi8v\nT7/85S/jd+QCADf5wj+2/sCBA7rtttsUCoXaPRQA4MQ1b948XXfddfxJAwBXS/ruKJ/m8/m0YMEC\n/dd//Vd75gEAAABc7wuPcOng/XG93uN6CQAAAOCEk/TlKCeddFLCdXkNDQ1qamrSb37zG1uCAQAA\nAG6V9Ah/9NFHEx536tRJ/fv3VzgcbvdQAAAAgJslPcJHjx5tZw4AAADghOGxLMtK5sAZM2YkdZuo\nJUuWHHeojqapqUlNTU1KsirHeb1exWIxp2MkxePxKC0tTS0tLfTbzujWXvRrL/q1V6r1S7f2SrV+\nc3JynI7RLpI+E56Tk6PFixdr4sSJKigo0M6dO/X8889r1qxZysvLszOj4zIyMlRbW5syt8nKzMxU\nY2Oj0zGSEggElJOTo/r6evptZ3RrL/q1F/3aK9X6pVt7pVq/bpH0CN+2bZtefPFFjRo1Kv7c2rVr\ntXDhQr388su2hAMAAADcKOn7C/7tb3/TWWedlfDc8OHD9frrr7d7KAAAAMDNkh7hQ4YM0Y9//OP4\nH1c0NjZqwYIFOuOMM2wLBwAAALhR0iP8kUce0Wuvvabs7Gx169ZN2dnZWrt2rRYvXmxnPgAAAMB1\nkr4mvE+fPlq3bp127dqlPXv2qEePHurdu7ed2QAAAABXOqafOf/JJ5/oL3/5i0pLS9W7d2/t2bNH\nu3fvtisbAAAA4EpJj/DS0lIVFRXpscce08KFCyVJZWVluuaaa2wLBwAAALhR0iP8uuuu05NPPqmV\nK1fK7z94Fcvw4cP15ptv2hYOAAAAcKOkR/iOHTs0btw4SYr/5My0tDS1tbXZkwwAAABwqaRH+KBB\ngw77oTx//vOfddppp7V7KAAAAMDNkr47yt13360JEybooosuUmNjo+bOnavnn39ey5YtszMfAAAA\n4DpJnwk/66yz9O677+rUU0/V7NmzdfLJJ+vNN9/UmWeeaWc+AAAAwHWSOhN+4MABjRs3Ti+//LJu\nuOEGuzMBAAAArpbUmXCfz6ft27crFovZnQcAAABwvaQvR7n55pt1zTXXqLy8XAcOHFAsFov/BQAA\nACB5SX9j5pw5cyRJS5Ysid+i0LIseTweHThwwJ50AAAAgAsddYR/+OGH6t69u7Zv324iDwAAAOB6\nRx3h/fv3VzQaVUFBgSTpkksu0bPPPmt7MAAAAMCtjjrCLctKePyXv/zFriwAkBKam5v18MMPq62t\nTePHj1d+fr7TkQAAKeaoI/zQ9d8AgIPmzJmj1atXS5J+9atf6ZVXXlH37t0dTgUASCVHHeFtbW1a\ns2ZN/Iz4px9L0tixY+1LCAAdSGNjY3yAS9Inn3yidevW6ZJLLnEwFQAg1Rx1hHft2lWzZ8+OP87L\ny0t47PF49MEHH9iTDgA6mIyMDEUiEVVWVsafO+mkkxxMBABIRUcd4Tt27DAQAwBSg8fj0cMPP6yb\nbrpJtbW1+uY3v6kzzzzT6VgAgBST9H3CAQAHDRkyRKtXr46fEW9tbXU6EgAgxST9EzMBAEDqq6qq\n0htvvJFwSRUA8xjhAACcILZs2aLRo0dr4sSJKiws1N///nenIwEnLEY4AAAniF//+teqqqqSJNXU\n1Oiee+5xOBFw4mKEAwBwgvB6vZ/7GIA5/NsHAMAJ4tprr1WPHj0kHbwF8Y033uhwIuDExd1RAAA4\nQfTt21evvvqq9u7dqzPOOENNTU3c3QdwiCtHeFlZmVauXCnLsjR06FCNHDky4eP79u3TH//4R+3d\nu1fjxo3TOeec41BSAADMysrK0oABAxQKhdTU1OR0HOCE5boRHovFtGLFCs2aNUuhUEiLFi1SUVGR\nIpFI/JjMzEyNHz9eW7ZscTApAAAATlSuuya8oqJCeXl5ysnJkc/nU3FxsbZu3ZpwTKdOndSzZ0++\nIQUAAACOcN0Kra2tVTgcjj8Oh8OKRqMOJgIAAAASue5ylOMVjUZVV1eX8FwwGJTfnzpV+Xw+BQIB\np2Mk5VCv9Nv+6NZe9Gsv+rVXqvVLt/ZKxX7dwD3v5P+EQiHV1NTEH0ej0YQz40ezYcMGlZaWJjw3\nevRojRkzpt0y4nC5ublOR3AturUX/dqLfu1Fv/ahWxyN60Z4fn6+qqqqVF1drWAwqE2bNmnKlClJ\nf/6wYcNUVFSU8FwwGNT+/fvV1tbW3nFtkZ6erubmZqdjJMXv9ys3N5d+bUC39qJfe9GvvVKtX7q1\nVyr26wauG+Fer1fjx4/X0qVLZVmWhgwZokgkovXr10uSSkpKVFdXp0WLFqm5uVkej0d/+9vfNG/e\nPKWnpyscDn/mmfPKysqUuZeq3+9PmayHtLW1pUzmVOuXbu1Fv/aiX3ulSr90a69U7NcNXDfCJamw\nsFCFhYUJz5WUlMR/HQwG9YMf/MB0LAAAAECSC++OAgAAAHR0jHAAAADAMFdejgIAdtuzZ48qKyuV\nl5fndBQAQAriTDgAHKOHHnpIQ4cO1amnnqorr7wyZe6AAADoOBjhAHAMWltbdeuttyoWi0mSVq1a\npT/96U8OpwIApBpGOAAcA8uyZFlWwnOHBjkAAMlihAPAMUhLS9NNN90Ufzxy5EhdcMEFDiYCAKQi\nvjETAI7Rt7/9bU2cOFE+n0/5+fmHnRkHAOBoGOEA8AX07dtXkUgkpX6aLgCg4+ByFAAAAMAwRjgA\nAABgGCMcAAAAMIwRDgAAABjGCAcAAAAMY4QDwBfQ0NCgffv2OR0DAJCiGOGAC3388ce68MILFQgE\ndPHFF6u6utrpSK7y7LPPqqioSJFIRPPmzeM+4QCAY8YIB1zotttu04YNG9TW1qbXXntNd999t9OR\nXKOtrU3XX3+9mpubJUlPP/20Vq1a5XAqAECqYYQDLvTpyyQqKysdSuI+Bw4ciA/wQ+rr6x1KAwBI\nVYxwwIWuuOIKeb0H//X2+/26/PLLHU7kHunp6Zo7d278cXFxsS644AIHEwEAUhE/th5woYkTJ6p3\n797avn27ioqKNHDgQKcjucoll1yiV199VXV1dZo+fbqysrKcjgQASDGMcMClSkpKdOGFF6qyslKt\nra1Ox3GVWbNmae/evZKk+fPnq7i4WEOGDHE4FQAglXA5CgAcg8bGxvgAlyTLsrR9+3YHEwEAUhEj\nHACOQWZmpkaMGBF/nJ2dreHDhzuYCACQirgcBQCO0SOPPKKHH35Yra2tmjRpkvLz852OBABIMYxw\nADhGWVlZuu666xSJRLjmHgDwhXA5CgAAAGAYIxwAAAAwjBEOAAAAGMYIBwAAAAxjhAMAAACGMcIB\nAAAAwxjhAAAAgGGMcAAAAMAwj2VZltMhOrqmpiY1NTUpVaryer2KxWJOx0iKx+NRWlqaWlpa6Led\n0a296Nde9GuvVOuXbu2Vav3m5OQ4HaNd8BMzk5CRkaHa2tqU+al4mZmZamxsdDpGUgKBgHJyclRf\nX0+/7Yxu7UW/9qJfe6Vav3Rrr1Tr1y24HAUAAAAwjBEOAAAAGMYIBwAAAAxjhAMAAACG8Y2ZcEws\nFtP27dt/fH0gAAAZCklEQVTl8/mcjgIAAGAUZ8LhiJaWFl155ZUaPny4evTooeeee87pSAAAAMYw\nwuGI5cuX69VXX5V0cJDfcMMNDicCAAAwhxEOR3z63qmpci9VAACA9sAIhyMmTpyoQYMGSTr406/m\nz5/vcCIAAABzGOFwRDAY1PLly/XHP/5RmzZt0ty5c52OBKCDuPfee1VYWKi+fftq3bp1TscBAFtw\ndxQ4JjMzU+ecc44ikYgqKyudjuM6GzZs0AcffKABAwZowIABTscBkrJu3TrdddddkqSamhrNnj1b\n7777rjwej8PJAKB9McIBF3rhhRd0zTXXKBaLye/3a/HixTrvvPOcjgUc1UcffZTwuKqqSk1NTcrM\nzHQoEQDYg8tRABf6wx/+oFgsJklqa2vTE0884XAiIDmjRo1SOByOPx49ejQDHIArMcIBF+rSpcvn\nPgY6qvLyckWj0fjjbdu2OZgGAOzDCAdcaMGCBRo2bJh8Pp/OOecc/fCHP3Q6kqvs2rVLc+bM0Ze/\n/GWtXLnS6TiuUl5envB47969amxsdCgNANiHa8IBF+rWrZteeukldenSRfv27eM+7O1s2rRp+uCD\nDyRJq1ev1ssvv6yBAwc6nModhg8frpycHFVXV0vichQA7sUIB1yMO0q0v4aGhvgAl6QDBw7ojTfe\nYIS3k/z8fC1btkzPPvusevbsqalTpzodCQBsweUoAHAMPut/bJqamhxI4l45OTnq1auXCgoKlJ6e\n7nQcALAFZ8IB4Bikp6crOztbNTU18ecGDx7sYCJ3qaqq0kUXXaTdu3dLkqZMmaL777/f4VQA0P44\nEw4Ax8Dr9eqhhx5S7969lZOToxtvvFFnn32207Fco7S0ND7AJemZZ57hGzMBuBJnwgHgGJ199tla\nv359/Ke98o2v7efTt9PMzs7mkhQArsSZcABAhzFq1Ch95zvfUUZGhrp3764HH3xQXi+/VQFwH/7L\nBgDoUObPn6+dO3dq7969Ou+885yOAwC2YIQDAAAAhjHCAQAAAMMY4QAAAIBhjHAAAADAMEY4AAAA\nYBgjHAAAADCMEQ4AAAAY5sqfmFlWVqaVK1fKsiwNHTpUI0eOPOyYFStW6L333lMgENDkyZPVo0cP\nB5ICAADgROS6M+GxWEwrVqzQjBkzNG/ePG3cuFGVlZUJx5SVlWn//v363ve+p4kTJ+qFF15wKC0A\nAABORK4b4RUVFcrLy1NOTo58Pp+Ki4u1devWhGO2bNmiwYMHS5J69eql5uZm1dXVOREXAAAAJyDX\njfDa2lqFw+H443A4rGg0+rnHhEKhw44BAAAA7OLKa8KPRzQaPeyseDAYlN+fOlX5fD4FAgGnYyTl\nUK/02/7o1l70ay/6tVeq9Uu39krFft3APe/k/4RCIdXU1MQfR6PRhLPeh4759zPf/37Mhg0bVFpa\nmnD86NGjNWbMGBtTIzc31+kIrkW39qJfe9GvvejXPnSLo3HdCM/Pz1dVVZWqq6sVDAa1adMmTZky\nJeGYoqIivfXWWyouLtauXbuUkZGhYDAoSRo2bJiKiooSjg8Gg9q/f7/a2tqMvY/jkZ6erubmZqdj\nJMXv9ys3N5d+bUC39qJfe9GvvVKtX7q1Vyr26wauG+Fer1fjx4/X0qVLZVmWhgwZokgkovXr10uS\nSkpK1L9/f5WVlen+++9XWlqaJk2aFP/8cDh82JlzSaqsrFRra6ux93E8/H5/ymQ9pK2tLWUyp1q/\ndGsv+rUX/dorVfqlW3ulYr9u4LoRLkmFhYUqLCxMeK6kpCTh8UUXXWQyEgAAABDnurujAAAAAB0d\nIxwAAAAwjBEOAAAAGMYIBwAAAAxjhAMAAACGMcIBAAAAwxjhAAAAgGGMcAAAAMAwRjgAAABgGCMc\nAAAAMIwRDgAAABjGCAcAAAAMY4QDAAAAhjHCAQAAAMMY4QAAAIBhjHAAAADAMEY4AAAAYBgjHAAA\nADCMEQ4AAAAYxggHAAAADGOEAwAAAIYxwgEAAADDGOEAAACAYYxwAAAAwDBGOAAAAGAYIxwAAAAw\njBEOAAAAGMYIBwAAAAxjhAMAAACGMcIBAAAAwxjhAAAAgGGMcAAAAMAwRjgAAABgmMeyLMvpEB1d\nU1OTmpqalCpVeb1exWIxp2MkxePxKC0tTS0tLfTbzujWXvRrL/q1V6r1S7f2SrV+c3JynI7RLvxO\nB0gFGRkZqq2tVWtrq9NRkpKZmanGxkanYyQlEAgoJydH9fX19NvO6NZe9Gsv+rVXqvVLt/ZKtX7d\ngstRAAAAAMMY4QAAAIBhjHAAAADAMEY4AAAAYBgjHAAAADCMEQ4AAAAYxggHAAAADGOEAwAAAIYx\nwgEAAADDGOEAAACAYYxwAAAAwDBGOAAAAGAYIxwAAAAwjBEOAAAAGMYIBwAAAAxjhAMAAACGMcIB\nAAAAwxjhAAAAgGGMcAAAAMAwRjgAAABgGCMcAAAAMIwRDgAAABjGCAcAAAAMY4QDAAAAhjHCAQAA\nAMMY4QAAAIBhjHAAAADAMEY4AAAAYBgjHAAAADCMEQ4AAAAYxggHAAAADPM7HaA9NTY26umnn1ZN\nTY1ycnJ02WWXKSMj47Djli1bpm3btqlTp0769re/7UBSAAAAnMhcdSZ87dq16tu3r7773e/q5JNP\n1l//+tfPPO6MM87Q9OnTDacDAAAADnLVCN+yZYvOOOMMSdLgwYO1ZcuWzzyuoKBAmZmZJqMBAAAA\nca4a4fX19QoGg5KkUCik+vp6hxMBAAAAh0u5a8KXLFmiurq6w54fO3bsYc95PJ5jfv1oNHrY6weD\nQfn9qVOVz+dTIBBwOkZSDvVKv+2Pbu1Fv/aiX3ulWr90a69U7NcNUu6dzJw584gfCwaDqqurUzAY\nVG1trTp16nTMr79hwwaVlpYmPFdQUKBLL71Uubm5x/x6+HzRaFRr1qzRsGHD6Led0a296Nde9Gsv\n+rUP3drr3/sNh8NOxzkurrocpaioSO+8844k6R//+IeKioqOeKxlWZ/5/LBhw/Stb30r/tfFF1+s\n8vLyzzz7juNXV1en0tJS+rUB3dqLfu1Fv/aiX/vQrb3c1G/KnQn/PCNGjNDTTz+tt99+W9nZ2brs\nssskSbW1tVq+fLmuvPJKSdIzzzyjHTt2qLGxUffcc4/GjBmjIUOGSJLC4XDK/58VAAAAOjZXjfCs\nrCzNmjXrsOdDoVB8gEvSlClTTMYCAAAAErjqchQAAAAgFfhuueWWW5wO0ZFZlqW0tDT16dNH6enp\nTsdxHfq1D93ai37tRb/2ol/70K293NSvxzrSdygiwSuvvKJt27bJ5/Opc+fOmjRpkjIyMpyO5Rr/\n/Oc/9Ze//EX79u3TN7/5TfXs2dPpSCmvrKxMK1eulGVZGjp0qEaOHOl0JNdYtmyZtm3bpk6dOunb\n3/6203FcpaamRs8995zq6+vl8Xg0dOhQnXXWWU7Hco22tjY9/PDDOnDggA4cOKCioiKdf/75Tsdy\nlVgspkWLFikcDmvatGlOx3Gde++9VxkZGfJ4PPJ6vfrWt77ldKQvzFXXhNupX79+Ov/88+X1evWn\nP/1Ja9eu5T9c7ahbt26aOnWqnn/+eaejuEIsFtOKFSs0a9YshUIhLVq0SEVFRYpEIk5Hc4UzzjhD\nX/rSl/Tcc885HcV1vF6vvvKVr6hHjx5qbm7WokWL1K9fP/7ZbSd+v1+zZs1SWlqaYrGYHnroIe3c\nuVO9e/d2OpprvPHGG4pEImpubnY6iit5PB59/etfd8VPPuea8CT169dPXu/Bunr16qVoNOpwInfp\n0qWL8vLynI7hGhUVFcrLy1NOTo58Pp+Ki4u1detWp2O5RkFBgSt+A+iIQqGQevToIUlKT09Xly5d\nVFtb63Aqd0lLS5N08Ky4ZVn8qW47qqmpUVlZmYYOHep0FFdzy0UcnAn/At5++20VFxc7HQM4otra\n2oRbbYbDYVVUVDiYCDh2+/fv14cffqj8/Hyno7jKocslqqqqVFJSoq5duzodyTVefvllXXDBBZwF\nt9mSJUvk9Xo1bNgwDRs2zOk4Xxgj/N8sWbLkM2/+Pm7cuPgP/nn11Vfl8/l0+umnm46X8pLpFwAk\nqbm5WU899ZQuvPDClP/mq47G6/Xq6quvVlNTk5YuXaodO3aoT58+TsdKeYe+T6RHjx7avn2703Fc\n6xvf+IZCoZDq6+u1ZMkSdenSRQUFBU7H+kIY4f9m5syZn/vxt99+W2VlZZ95L3Ic3dH6RfsJhUKq\nqamJP45Go/wQKqSMAwcO6KmnntLgwYM1YMAAp+O4VkZGhvr37689e/YwwtvBzp07tXXrVpWVlamt\nrU3Nzc169tlndckllzgdzVVCoZAkqVOnTho4cKAqKioY4W5XVlamdevW6aqrrpLfT23o2PLz81VV\nVaXq6moFg0Ft2rSJH1LVztxyTWJHtGzZMkUiEe6KYoP6+nr5fD5lZGSotbVV77//vs477zynY7nC\n+eefH79hw44dO7Ru3ToGeDtraWmRZVlKT09XS0uL3n//fY0ePdrpWF8YtyhM0gMPPKADBw7Evxmr\nV69emjBhgsOp3ONf//qXXnrpJTU0NCgjI0Pdu3fX9OnTnY6V0v79FoVDhgzRqFGjnI7kGs8884x2\n7NihxsZGderUSWPGjNGQIUOcjuUKO3fu1MMPP6yuXbvK4/FIOnjJWmFhocPJ3OGjjz6K39XHsiyd\nfvrpGjFihMOp3OfQCOcWhe1r//79euKJJ+TxeBSLxXTaaael9O9tjHAAAADAMG5RCAAAABjGCAcA\nAAAMY4QDAAAAhjHCAQAAAMMY4QAAAIBhjHAAAADAMEY4AAAAYBgjHAAAADCMEQ4AAAAYxggHAAAA\nDGOEAwAAAIYxwgEAAADDGOEAAACAYYxwAAAAwDBGOAAAAGAYIxwAAAAwjBEOAAAAGMYIBwAAAAxj\nhAMAAACGMcIBdEh9+vRRVlaWsrOz1blzZ40cOVK//e1vZVmW09GSdvLJJ2v16tVOx/hcixcv1qhR\no5yOYZvy8nJ5vV7FYjGnowBAAkY4gA7J4/HoxRdfVE1NjcrLy3XTTTfpzjvv1De+8Q2no7mKZVny\neDxOx7DNofeXSv/zBuDEwAgH0GEdGk6hUEgTJkzQk08+qcWLF2vz5s2SpBUrVmjo0KHKzs5WQUGB\nfv7zn8c/d8KECfr1r3+d8HqDBw/WsmXLDvs6h86WPvjgg8rPz1d+fr7uvvvu+MdbWlp03XXXKT8/\nX7169dL3v/99tba2SpI++eQTTZw4Ubm5ucrLy9Po0aMlSTNnztTOnTs1ceJEhcNh3XXXXYd93SN9\n7l133aUpU6YkHPu9731P3//+9yVJY8aM0U9/+lONGDFCoVBIkyZN0ieffKLp06crOztbw4cP186d\nO+Of6/V69atf/Ur9+vVT165ddcMNN0iStmzZomuuuUavv/66QqGQOnfuLEmKRqOaOXOmunbtqpNP\nPlm33357/LUWL16skSNH6gc/+IFyc3NVWFiodevW6ZFHHlHv3r3VvXt3LVmy5Ih/T/fu3atJkyYp\nLy9P/fv31+9+97v4x37+85/r8ssv16xZsxQOh3Xaaafp73//e/zjd955p3r16qVwOKyBAwdqzZo1\nkg7+c/KLX/xCp5xyiiKRiKZOnarq6mpJineak5OjcDisN95444jZAMAoCwA6oD59+lirVq067Pne\nvXtb//M//2NZlmWVlpZamzZtsizLsjZu3Gh1797dWrZsmWVZlvXUU09Zw4cPj3/eO++8Y3Xp0sVq\nbW097DV37NhheTwea9q0aVZjY6O1ceNGKxKJxL/+T3/6U+vss8+29u3bZ+3bt88655xzrJ/97GeW\nZVnW/PnzrWuuucY6cOCA1dbWZq1duzbhPaxevfqI7/FIn7t3714rGAxaNTU1lmVZVltbm9W1a1fr\n7bfftizLss477zyrsLDQ2r59uxWNRq1BgwZZhYWF1urVq60DBw5YM2fOtGbPnh3/Oh6Pxxo7dqxV\nXV1t7dq1y+rfv7/10EMPWZZlWY888og1atSohFwzZsywJk+ebNXX11s7duyw+vfvb/3+97+PHx8I\nBKzFixdbsVjM+slPfmL16tXL+s53vmO1tLRYr7zyihUKhaz6+vrPfM+jRo2KH/vOO+9YkUjEWrNm\njWVZlnXLLbdYmZmZ1sqVK61YLGbNnz/fOuussyzLsqytW7daJ510kvXhhx9almVZ5eXl1gcffGBZ\nlmXdd9991tlnn23t2bPHamlpsa6++mrriiuuiP+99Xq9ViwWO+LfBwBwAmfCAaSUnj17qqqqSpJ0\n7rnn6tRTT5UkFRcXa+rUqSotLZUkffWrX1VZWZnef/99SdKjjz6qyy+/XH6//4ivfcsttygjI0PF\nxcW66qqr9Pjjj0uS/vCHP+jmm29WXl6e8vLydPPNN2vp0qWSpEAgoL1792r79u3y+XwaMWJEwmta\nn3MZxJE+t3v37ho1apSefvppSdJLL72kSCSiM844I/65V111lfr06aNQKKQLL7xQhYWFGjNmjLxe\nry677DK9/fbbCV/rpptuUnZ2tnr16qXrrrsu/t4+LRaL6cknn9QvfvELZWVlqaCgQD/84Q/j71c6\neK37zJkz5fF4dPnll2vPnj26+eabFQgEdMEFFygtLU3vvffeYa+9e/duvf7667rzzjsVCAQ0ePBg\nzZkzJ+HM+ciRI/WVr3xFHo9HM2bM0LvvvitJ8vl8amlp0aZNm9TW1qbevXvr5JNPliT99re/1e23\n364ePXooEAjoZz/7mZ555hnFYrF4/5/39wEAnMAIB5BSKioq4pdNvPHGGxo7dqy6du2qnJwc/fa3\nv9W+ffskSenp6fra176mRx99VJZl6fHHH9eMGTOO+Loej0e9evWKPy4oKNCePXskSXv27FHv3r0/\n82PXX3+9+vXrpy9/+cs65ZRTdOeddyb9Xm644YYjfu7MmTP16KOPSpIee+yxw7J369Yt/uvMzMzD\nHtfV1SUcf6T39mn79u2Lj9x/P76iouKIX1uSunTp8rlfXzrYY+fOnZWVlXXE1+7evXv811lZWWpq\nalIsFlO/fv1033336ZZbblG3bt00bdo0ffjhh5IOXk508cUXq3PnzurcubMGDRqkQCCgjz76yNXX\nuwNIbYxwACnjrbfe0p49e+J387jyyis1efJkVVRUqLq6WnPnzk0443loyK5atUqdOnXS8OHDj/ja\nlmVp165d8cc7d+5Uz549JR08+15eXh7/WHl5efxjwWBQd911l95//30tX75c99xzT/xa5aMNwE6d\nOh3xcydPnqx3331X//znP/XCCy/oyiuvPJaqDnOk9/bpjF26dFEgEDjs/ebn5x/X15f+/59i1NfX\nJ2RJ9rWnTp2qv/71r/FsN954oySpd+/eeumll1RVVaWqqirt379f9fX16tGjByMcQIfFCAfQ4dXW\n1uqFF17QFVdcoRkzZmjQoEGSpLq6OuXm5ioQCOjNN9/UH/7wh4TPO/vss+XxePTDH/7wc8+CH7Jw\n4UI1Njbqn//8px5++GFNnTpVknTFFVfotttu0759+7Rv3z4tXLgw/novvvhi/JKXUCgkv98vn88n\n6eAZ4w8++OCIX++zPtfrPfif5YyMDF1yySWaNm2ahg8fnnAm+4v45S9/qerqau3atUv3339//L11\n69ZNu3fvjn+jqdfr1de+9jUtWLBAdXV1Ki8v17333vu5/SV7qUevXr10zjnnaP78+Wpubta7776r\nhx56KKnX3rZtm9asWaOWlhalpaUpMzMz3tXcuXP14x//OP7NqJWVlVq+fLkkKRKJyOv1xnsGgI6C\nEQ6gw5o4caKys7PVu3dv3XHHHfrRj36k3//+9/GP/+Y3v9FPf/pTZWdn67bbbtPll19+2GvMnDlT\nmzZt0vTp04/69UaPHq1TTjlFF1xwgW644QaNGzdOkvSTn/xEJSUlOv300zV48GCVlJRowYIFkqSy\nsjKdf/75CoVCGjFihObNm6dzzz1XkjR//nwtXLhQnTt31j333HPY1/uszz10Nw9JmjVrljZu3KiZ\nM2cmfN4XObs7adIkDRs2TEOHDtXEiRM1e/ZsSdLYsWN16qmnqnv37uratask6YEHHlBWVpb69u2r\nc889V9OnT9dVV111xNf+dJ7Py/f4449r+/bt6tmzpy699FItXLhQY8aMOeprNzc366abblIkElHP\nnj1VWVmpO+64Q5J07bXXatKkSfryl7+s7OxsnXPOOXrzzTclHbw0ZsGCBRoxYoQ6d+4cfx4AnOax\n+G4VAC726KOPatGiRXr11VePeEx5ebn69u2r1tbW+NnVjmD37t0aMGCAPvzwQwWDwS/8Ol6vV++9\n95769u3bjukAAMej4/xuAwDtrKGhQb/+9a81d+7cox7b0c5HxGIx3XXXXZo6depxDXAAQMfECAfg\nSq+88oq6du2qHj166Iorrjjq8R3pG/gaGhqUnZ2t1atXJ/wAoi+qI703AMBBXI4CAAAAGMaZcAAA\nAMAwRjgAAABgGCMcAAAAMIwRDgAAABjGCAcAAAAMY4QDAAAAhv0/mOqbmuvpaB8AAAAASUVORK5C\nYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10ccd4b00>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<ggplot: (-9223372036578223160)>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ggplot(aes(x='DPI', y='freq.var'), data=AA) +\\\n",
    "    geom_point()  + ylab(\"Frequency\")+xlab(\"Day post symptom onset\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18\n"
     ]
    }
   ],
   "source": [
    "AA_nonsynom_n=len(AA)\n",
    "print(AA_nonsynom_n)\n",
    "write_to_summary(\"Antigenic iSNV n:\",AA_nonsynom_n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "HA_nonsyn_n=len(HA_nonsyn)\n",
    "write_to_summary(\"NS HA n:\",HA_nonsyn_n)\n",
    "#HA_nonsyn_n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proportion of snv in antigenic sites 0.025245 - 0.009818\n",
      "Propotion of infections with antigenic sites 0.075000 - 0.035000 \n"
     ]
    }
   ],
   "source": [
    "frac_isnv = \"Proportion of snv in antigenic sites %f - %f\" % (len(AA)/len(snv), (len(AA.loc[AA[\"Source\"]==\"Wiley\"])+len(AA.loc[AA[\"Source\"]==\"Smith\"])+len(AA.loc[AA[\"Source\"]==\"Caton\"]))/len(snv))\n",
    "\n",
    "print(frac_isnv)\n",
    "\n",
    "better = AA.loc[(AA[\"Source\"]==\"Wiley\")|(AA[\"Source\"]==\"Smith\")|(AA[\"Source\"]==\"Caton\")]\n",
    "frac_infection = \"Propotion of infections with antigenic sites %f - %f \" % (len(AA.ENROLLID.unique())/200,len(better.ENROLLID.unique())/200)\n",
    "\n",
    "\n",
    "print(frac_infection)\n",
    "\n",
    "write_to_summary(\"Antigenic iSNV %:\",frac_isnv)\n",
    "write_to_summary(\"Antigenic infection n:\",frac_infection)\n",
    "#len(AA)/len(snv.loc[snv[\"class_factor\"]==\"Nonsynonymous\"])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Amino acids for Nextflu\n",
    "\n",
    "\n",
    "This a little old code but it used regex to get the amino acid position and and the variant or ref amino acid.\n",
    "\n",
    "H3_name is the minor\n",
    "H3_name_major is the major amino acid."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "H3=[]\n",
    "H1=[]\n",
    "H3_major= []\n",
    "H1_major = []\n",
    "pat = re.compile('\\w(\\d+) (in|of) (HA\\d)') \n",
    "for index, row in HA_nonsyn.iterrows():\n",
    "    H3_pos = pat.match(row.loc[\"PDB_4HMG\"])\n",
    "    H1_pos = pat.match(row.loc[\"PDB_4JTV\"])\n",
    "    var = row.loc[\"Var_AA\"].split(\"'\")[1]\n",
    "    ref = row.loc[\"Ref_AA\"].split(\"'\")[1]\n",
    "    if H3_pos:\n",
    "        H3.append(H3_pos.group(3)+\":\"+H3_pos.group(1)+var)\n",
    "        H3_major.append(H3_pos.group(3)+\":\"+H3_pos.group(1)+ref)\n",
    "    else:\n",
    "        H3.append(\"-\")\n",
    "        H3_major.append(\"-\")\n",
    "    if H1_pos:\n",
    "        H1.append(H1_pos.group(3)+\":\"+H1_pos.group(1)+var)\n",
    "        H1_major.append(H1_pos.group(3)+\":\"+H1_pos.group(1)+ref)\n",
    "    else:\n",
    "        H1.append(\"-\")\n",
    "        H1_major.append(\"-\")\n",
    "        \n",
    "        \n",
    "HA_nonsyn[\"H3_name\"] = H3\n",
    "HA_nonsyn[\"H3_name_major\"] = H3_major\n",
    "\n",
    "\n",
    "HA_nonsyn[\"H1_name\"] = H1\n",
    "HA_nonsyn[\"H1_name_major\"] = H1_major\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "HA_nonsyn.to_csv(\"./minor_nonsynom.cvs\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Is  AA_pos  base 0 or base 1\n",
    "\n",
    "I expect a G at PB1 594 (base1) )(593-base 0)  in sample MH7800 HK_8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for seq_record in SeqIO.parse(\"../data/processed/HK_8/coding_fa/MH7800_A.removed.parsed.coding.fasta\", \"fasta\"):    \n",
    "        if seq_record.name == \"PB1\":\n",
    "            PB1=seq_record\n",
    "            PB1_prot = seq_record.seq.translate(to_stop=True)\n",
    "        if seq_record.name == \"NR\":\n",
    "            NR=seq_record\n",
    "            NR_prot = seq_record.seq.translate(to_stop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MDVNPTLLFLKVPAQNAISTTFPYTGDPPYSHGTGTGYTMDTVNRTHQYSERGK...RQK', ExtendedIUPACProtein())"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PB1_prot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'G'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PB1_prot[593]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'N'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NR_prot[160]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is is base 1."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sci-py3",
   "language": "python",
   "name": "sci-py3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
